--- a/net/minecraft/entity/projectile/FishingBobberEntity.java
+++ b/net/minecraft/entity/projectile/FishingBobberEntity.java
@@ -3,6 +3,11 @@
 import java.util.Collections;
 import java.util.List;
 import javax.annotation.Nullable;
+
+import org.vivecraft.api.NetworkHelper;
+import org.vivecraft.api.ServerVivePlayer;
+import org.vivecraft.provider.MCOpenVR;
+
 import net.minecraft.advancements.CriteriaTriggers;
 import net.minecraft.block.Block;
 import net.minecraft.block.Blocks;
@@ -26,6 +31,7 @@
 import net.minecraft.stats.Stats;
 import net.minecraft.tags.FluidTags;
 import net.minecraft.tags.ItemTags;
+import net.minecraft.util.Hand;
 import net.minecraft.util.SoundEvents;
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.util.math.EntityRayTraceResult;
@@ -85,10 +91,26 @@
         float f3 = MathHelper.sin(-f1 * ((float)Math.PI / 180F) - (float)Math.PI);
         float f4 = -MathHelper.cos(-f * ((float)Math.PI / 180F));
         float f5 = MathHelper.sin(-f * ((float)Math.PI / 180F));
-        double d0 = this.angler.getPosX() - (double)f3 * 0.3D;
-        double d1 = this.angler.getPosYEye();
-        double d2 = this.angler.getPosZ() - (double)f2 * 0.3D;
-        this.setLocationAndAngles(d0, d1, d2, f1, f);
+       	//VIVE USE CONTROLLER  	
+    	ServerVivePlayer vp = NetworkHelper.vivePlayers.get(angler.getUniqueID());
+    	if (vp !=null && vp.isVR())
+    	{
+			int c = 0;
+			if (p_i4184_1_.getActiveHand() == Hand.OFF_HAND) c = 1;
+    		Vec3d aim = vp.getControllerDir(c);
+    		Vec3d aimSource = vp.getControllerPos(vp.isSeated() ? 0 : c);
+    		float pitch = (float)Math.toDegrees(Math.asin(aim.y/aim.length()));
+    		float yaw = (float)Math.toDegrees(Math.atan2(aim.x, -aim.z));   
+    		this.setLocationAndAngles(aimSource.x + aim.x*0.6f, aimSource.y+aim.y*0.6f, aimSource.z+aim.z*0.6f, yaw, pitch);
+    		f = pitch;
+    		f1 = yaw;
+    	}else {
+	        double d0 = this.angler.getPosX() - (double)f3 * 0.3D;
+	        double d1 = this.angler.getPosYEye();
+	        double d2 = this.angler.getPosZ() - (double)f2 * 0.3D;
+	        this.setLocationAndAngles(d0, d1, d2, f1, f);
+    	}
+    	//END VIVE
         Vec3d vec3d = new Vec3d((double)(-f3), (double)MathHelper.clamp(-(f5 / f4), -5.0F, 5.0F), (double)(-f2));
         double d3 = vec3d.length();
         vec3d = vec3d.mul(0.6D / d3 + 0.5D + this.rand.nextGaussian() * 0.0045D, 0.6D / d3 + 0.5D + this.rand.nextGaussian() * 0.0045D, 0.6D / d3 + 0.5D + this.rand.nextGaussian() * 0.0045D);
@@ -236,6 +258,20 @@
             this.setMotion(this.getMotion().scale(0.92D));
             this.recenterBoundingBox();
         }
+        
+        if(angler.isUser()) {
+        	//Vivecraft nibblehaptics
+        	ServerVivePlayer vp = NetworkHelper.vivePlayers.get(angler.getUniqueID());
+        	if (vp !=null && vp.isVR()) {
+        		if(this.currentState == State.BOBBING) {
+        			if(this.getMotion().y <= -0.24) { //server makes the bobber jump downwards on a bite.
+        				MCOpenVR.triggerHapticPulse(0,250); //VIVECRAFT go thump.
+        			}				
+        		}
+        	}
+        }
+    	//
+    	
     }
 
     private boolean shouldStopFishing()
