--- a/net/minecraft/entity/projectile/ProjectileEntity.java
+++ b/net/minecraft/entity/projectile/ProjectileEntity.java
@@ -2,9 +2,14 @@
 
 import java.util.UUID;
 import javax.annotation.Nullable;
+
+import org.vivecraft.api.NetworkHelper;
+import org.vivecraft.api.ServerVivePlayer;
+
 import net.minecraft.block.BlockState;
 import net.minecraft.entity.Entity;
 import net.minecraft.entity.EntityType;
+import net.minecraft.entity.player.PlayerEntity;
 import net.minecraft.nbt.CompoundNBT;
 import net.minecraft.util.math.BlockRayTraceResult;
 import net.minecraft.util.math.EntityRayTraceResult;
@@ -114,6 +119,27 @@
 
     public void func_234612_a_(Entity p_234612_1_, float p_234612_2_, float p_234612_3_, float p_234612_4_, float p_234612_5_, float p_234612_6_)
     {
+    	//VIVE SET AIM AND SPEED
+    	//Crossbow aim is not done here. See crossbowitem.
+    	ServerVivePlayer vp = NetworkHelper.vivePlayers.get(p_234612_1_.getUniqueID());
+    	if (vp !=null && vp.isVR())
+    	{ 
+    		//Trident and throwables
+    		Vector3d aim = vp.getControllerDir(vp.activeHand);
+    		
+    		//Roomscale Bow
+    		if (this instanceof AbstractArrowEntity && !(this instanceof TridentEntity)) {
+				if( !vp.isSeated() && vp.getDraw() > 0){ 
+					aim = vp.getControllerPos(1,(PlayerEntity) p_234612_1_).subtract(vp.getControllerPos(0,(PlayerEntity) p_234612_1_)).normalize();
+					p_234612_5_ = p_234612_5_ * vp.getDraw();  
+					((AbstractArrowEntity)this).setDamage(((AbstractArrowEntity)this).getDamage()*2);
+				}
+    		}
+			
+			p_234612_2_ = -(float)Math.toDegrees(Math.asin(aim.y/aim.length()));
+			p_234612_3_ = (float)Math.toDegrees(Math.atan2(-aim.x, aim.z));   
+    	}
+    	///
         float f = -MathHelper.sin(p_234612_3_ * ((float)Math.PI / 180F)) * MathHelper.cos(p_234612_2_ * ((float)Math.PI / 180F));
         float f1 = -MathHelper.sin((p_234612_2_ + p_234612_4_) * ((float)Math.PI / 180F));
         float f2 = MathHelper.cos(p_234612_3_ * ((float)Math.PI / 180F)) * MathHelper.cos(p_234612_2_ * ((float)Math.PI / 180F));
