--- a/net/minecraft/client/AbstractOption.java
+++ b/net/minecraft/client/AbstractOption.java
@@ -151,7 +151,7 @@
     });
     public static final SliderPercentageOption FRAMERATE_LIMIT = new SliderPercentageOption("options.framerateLimit", 0.0D, 260.0D, 5.0F, (p_lambda$static$27_0_) ->
     {
-        return p_lambda$static$27_0_.vsync ? FRAMERATE_LIMIT.getMinValue() : (double)p_lambda$static$27_0_.framerateLimit;
+        return p_lambda$static$27_0_.vsync ? 0 : (double)p_lambda$static$27_0_.framerateLimit;
     }, (p_lambda$static$28_0_, p_lambda$static$28_1_) ->
     {
         p_lambda$static$28_0_.framerateLimit = (int)p_lambda$static$28_1_.doubleValue();
@@ -159,7 +159,7 @@
 
         if (p_lambda$static$28_0_.framerateLimit <= 0)
         {
-            p_lambda$static$28_0_.framerateLimit = (int)FRAMERATE_LIMIT.getMaxValue();
+            p_lambda$static$28_0_.framerateLimit = 260;
             p_lambda$static$28_0_.vsync = true;
         }
 
