--- a/net/minecraft/world/biome/Biome.java
+++ b/net/minecraft/world/biome/Biome.java
@@ -820,7 +820,7 @@
         NETHER("nether");
 
         public static final Codec<Biome.Category> field_235102_r_ = IStringSerializable.func_233023_a_(Biome.Category::values, Biome.Category::func_235103_a_);
-        private static final Map<String, Biome.Category> BY_NAME = Arrays.stream(values()).collect(Collectors.toMap(Biome.Category::getName, (p_222353_0_) -> {
+        private static final Map<String, Biome.Category> BY_NAME = Arrays.stream(values()).collect(Collectors.toMap(Biome.Category::getName1, (p_222353_0_) -> {
             return p_222353_0_;
         }));
         private final String name;
@@ -830,11 +830,6 @@
             this.name = p_i4018_3_;
         }
 
-        public String getName()
-        {
-            return this.name;
-        }
-
         public static Biome.Category func_235103_a_(String p_235103_0_)
         {
             return BY_NAME.get(p_235103_0_);
@@ -875,7 +870,7 @@
         SNOW("snow");
 
         public static final Codec<Biome.RainType> field_235121_d_ = IStringSerializable.func_233023_a_(Biome.RainType::values, Biome.RainType::func_235122_a_);
-        private static final Map<String, Biome.RainType> BY_NAME = Arrays.stream(values()).collect(Collectors.toMap(Biome.RainType::getName, (p_222360_0_) -> {
+        private static final Map<String, Biome.RainType> BY_NAME = Arrays.stream(values()).collect(Collectors.toMap(Biome.RainType::getName1, (p_222360_0_) -> {
             return p_222360_0_;
         }));
         private final String name;
@@ -885,11 +880,6 @@
             this.name = p_i1934_3_;
         }
 
-        public String getName()
-        {
-            return this.name;
-        }
-
         public static Biome.RainType func_235122_a_(String p_235122_0_)
         {
             return BY_NAME.get(p_235122_0_);
