--- a/net/optifine/gui/GuiMessage.java
+++ b/net/optifine/gui/GuiMessage.java
@@ -62,7 +62,7 @@
     {
         this.ticksUntilEnable = ticksUntilEnable;
 
-        for (Button button : this.buttonList)
+        for (Widget button : this.buttonList)
         {
             button.active = false;
         }
@@ -74,7 +74,7 @@
 
         if (--this.ticksUntilEnable == 0)
         {
-            for (Button button : this.buttonList)
+            for (Widget button : this.buttonList)
             {
                 button.active = true;
             }
