--- a/net/minecraft/world/dimension/EndDimension.java
+++ b/net/minecraft/world/dimension/EndDimension.java
@@ -15,6 +15,7 @@
 import net.minecraft.world.gen.ChunkGeneratorType;
 import net.minecraft.world.gen.EndGenerationSettings;
 import net.minecraft.world.server.ServerWorld;
+import net.optifine.reflect.Reflector;
 
 public class EndDimension extends Dimension
 {
@@ -24,8 +25,17 @@
     public EndDimension(World p_i1397_1_, DimensionType p_i1397_2_)
     {
         super(p_i1397_1_, p_i1397_2_, 0.0F);
-        CompoundNBT compoundnbt = p_i1397_1_.getWorldInfo().getDimensionData(DimensionType.THE_END);
-        this.dragonFightManager = p_i1397_1_ instanceof ServerWorld ? new DragonFightManager((ServerWorld)p_i1397_1_, compoundnbt.getCompound("DragonFight")) : null;
+        if (p_i1397_1_ != null) {
+   	     CompoundNBT compoundnbt = p_i1397_1_.getWorldInfo().getDimensionData(p_i1397_2_);
+   	     
+   	     if(Reflector.DragonFightManager_Constructor.exists())
+   	    	 this.dragonFightManager = p_i1397_1_ instanceof ServerWorld ? (DragonFightManager) Reflector.newInstance(Reflector.DragonFightManager_Constructor, p_i1397_1_, compoundnbt.getCompound("DragonFight"), this) : null;
+   	     else
+   	    	 this.dragonFightManager = p_i1397_1_ instanceof ServerWorld ? new DragonFightManager((ServerWorld)p_i1397_1_, compoundnbt.getCompound("DragonFight")) : null;
+
+        } else {
+            this.dragonFightManager = null;
+        }
     }
 
     public ChunkGenerator<?> createChunkGenerator()
@@ -119,8 +129,11 @@
         {
             compoundnbt.put("DragonFight", this.dragonFightManager.write());
         }
-
-        this.world.getWorldInfo().setDimensionData(DimensionType.THE_END, compoundnbt);
+        
+        if(Reflector.DragonFightManager_Constructor.exists())
+        	this.world.getWorldInfo().setDimensionData(this.world.getDimension().getType(), compoundnbt);
+        else
+        	this.world.getWorldInfo().setDimensionData(DimensionType.THE_END, compoundnbt);
     }
 
     public void tick()
