--- a/net/minecraft/world/dimension/EndDimension.java
+++ b/net/minecraft/world/dimension/EndDimension.java
@@ -24,8 +24,12 @@
     public EndDimension(World p_i1397_1_, DimensionType p_i1397_2_)
     {
         super(p_i1397_1_, p_i1397_2_, 0.0F);
-        CompoundNBT compoundnbt = p_i1397_1_.getWorldInfo().getDimensionData(DimensionType.THE_END);
-        this.dragonFightManager = p_i1397_1_ instanceof ServerWorld ? new DragonFightManager((ServerWorld)p_i1397_1_, compoundnbt.getCompound("DragonFight")) : null;
+        if (p_i1397_1_ != null) {
+   	     CompoundNBT compoundnbt = p_i1397_1_.getWorldInfo().getDimensionData(p_i1397_2_);
+   	     this.dragonFightManager = p_i1397_1_ instanceof ServerWorld ? new DragonFightManager((ServerWorld)p_i1397_1_, compoundnbt.getCompound("DragonFight")) : null;
+        } else {
+            this.dragonFightManager = null;
+        }
     }
 
     public ChunkGenerator<?> createChunkGenerator()
