--- a/net/minecraft/client/renderer/entity/EntityRendererManager.java
+++ b/net/minecraft/client/renderer/entity/EntityRendererManager.java
@@ -55,7 +55,7 @@
 public class EntityRendererManager
 {
     private static final RenderType SHADOW_RENDER_TYPE = RenderType.getEntityNoOutline(new ResourceLocation("textures/misc/shadow.png"));
-    private final Map < EntityType<?>, EntityRenderer<? >> renderers = Maps.newHashMap();
+    private final Map<EntityType, EntityRenderer> renderers = Maps.newHashMap();
     private final Map<String, PlayerRenderer> skinMap = Maps.newHashMap();
     private final PlayerRenderer playerRenderer;
     private final FontRenderer textRenderer;
