--- a/net/optifine/shaders/gui/GuiShaders.java
+++ b/net/optifine/shaders/gui/GuiShaders.java
@@ -37,10 +37,10 @@
     public static final int EnumOS_SOLARIS = 3;
     public static final int EnumOS_LINUX = 4;
 
-    public GuiShaders(Screen par1GuiScreen, GameSettings par2GameSettings)
+    public GuiShaders(Screen par1Screen, GameSettings par2GameSettings)
     {
         super(new StringTextComponent(I18n.format("of.options.shadersTitle")));
-        this.parentGui = par1GuiScreen;
+        this.parentGui = par1Screen;
     }
 
     public void init()
