--- a/net/minecraft/client/renderer/vertex/DefaultVertexFormats.java
+++ b/net/minecraft/client/renderer/vertex/DefaultVertexFormats.java
@@ -15,7 +15,8 @@
     public static final VertexFormatElement PADDING_1B = new VertexFormatElement(0, VertexFormatElement.Type.BYTE, VertexFormatElement.Usage.PADDING, 1);
     public static VertexFormat BLOCK = new VertexFormat(ImmutableList.<VertexFormatElement>builder().add(POSITION_3F).add(COLOR_4UB).add(TEX_2F).add(TEX_2SB).add(NORMAL_3B).add(PADDING_1B).build());
     public static VertexFormat ENTITY = new VertexFormat(ImmutableList.<VertexFormatElement>builder().add(POSITION_3F).add(COLOR_4UB).add(TEX_2F).add(TEX_2S).add(TEX_2SB).add(NORMAL_3B).add(PADDING_1B).build());
-    private static final VertexFormat BLOCK_VANILLA = BLOCK;
+    private static final VertexFormat ENTITY_VANILLA = ENTITY;
+    public static final VertexFormat BLOCK_VANILLA = BLOCK; // VIVECRAFT: meh just make it public
     private static final VertexFormat BLOCK_SHADERS = SVertexFormat.makeExtendedFormatBlock(BLOCK_VANILLA);
     public static final int BLOCK_VANILLA_SIZE = BLOCK_VANILLA.getSize();
     public static final int BLOCK_SHADERS_SIZE = BLOCK_SHADERS.getSize();
@@ -38,6 +39,12 @@
     @Deprecated
     public static final VertexFormat POSITION_TEX_COLOR_NORMAL = new VertexFormat(ImmutableList.<VertexFormatElement>builder().add(POSITION_3F).add(TEX_2F).add(COLOR_4UB).add(NORMAL_3B).add(PADDING_1B).build());
 
+    //Vivecraft
+    public static final VertexFormat POSITION_TEX_LMAP_COLOR_NORMAL = new VertexFormat(ImmutableList.<VertexFormatElement>builder().add(POSITION_3F).add(TEX_2F).add(TEX_2SB).add(COLOR_4UB).add(NORMAL_3B).add(PADDING_1B).build());
+    public static final VertexFormat POSITION_COLOR_NORMAL = new VertexFormat(ImmutableList.<VertexFormatElement>builder().add(POSITION_3F).add(COLOR_4UB).add(NORMAL_3B).add(PADDING_1B).build());
+    ///
+    
+    
     public static void updateVertexFormats()
     {
         if (Config.isShaders())
