--- a/net/minecraft/client/KeyboardListener.java
+++ b/net/minecraft/client/KeyboardListener.java
@@ -2,6 +2,12 @@
 
 import java.util.Locale;
 import javax.annotation.Nullable;
+
+import org.lwjgl.glfw.GLFW;
+import org.vivecraft.gameplay.screenhandlers.KeyboardHandler;
+import org.vivecraft.gameplay.screenhandlers.RadialHandler;
+import org.vivecraft.settings.VRHotkeys;
+
 import net.minecraft.block.BlockState;
 import net.minecraft.client.gui.IGuiEventListener;
 import net.minecraft.client.gui.INestedGuiEventHandler;
@@ -339,6 +345,23 @@
     {
         if (windowPointer == this.mc.getMainWindow().getHandle())
         {
+            //Vivecraft esc
+            if (key == GLFW.GLFW_KEY_ESCAPE && action == GLFW.GLFW_PRESS)
+            {                       
+            	if(KeyboardHandler.Showing) {
+            		KeyboardHandler.setOverlayShowing(false);
+            		return;
+            	}
+            	if(RadialHandler.isShowing()) {
+            		RadialHandler.setOverlayShowing(false, null);
+            		return; 		
+            	}
+            }
+
+            if (VRHotkeys.handleKeyboardInputs(key, scanCode, action, modifiers))
+                return;
+            ///
+            
             if (this.debugCrashKeyPressTime > 0L)
             {
                 if (!InputMappings.isKeyDown(Minecraft.getInstance().getMainWindow().getHandle(), 67) || !InputMappings.isKeyDown(Minecraft.getInstance().getMainWindow().getHandle(), 292))
@@ -373,13 +396,14 @@
                         if (Screen.hasControlDown())
                         {
                         }
-
+                        this.mc.grabScreenShot = true;
+                        /* VIVECRAFT
                         ScreenShotHelper.saveScreenshot(this.mc.gameDir, this.mc.getMainWindow().getFramebufferWidth(), this.mc.getMainWindow().getFramebufferHeight(), this.mc.getFramebuffer(), (p_lambda$onKeyEvent$3_1_) ->
                         {
                             this.mc.execute(() -> {
                                 this.mc.ingameGUI.getChatGUI().printChatMessage(p_lambda$onKeyEvent$3_1_);
                             });
-                        });
+                        });*/
                         return;
                     }
                 }
@@ -453,8 +477,9 @@
                     return;
                 }
             }
-
-            if (this.mc.currentScreen == null || this.mc.currentScreen.passEvents)
+            
+            // VICECRAFT: Stop shit happening when VRHotkeys is in use
+            if ((this.mc.currentScreen == null || this.mc.currentScreen.passEvents) && !InputMappings.isKeyDown(GLFW.GLFW_KEY_RIGHT_CONTROL))
             {
                 InputMappings.Input inputmappings$input = InputMappings.getInputByCode(key, scanCode);
 
@@ -512,6 +537,7 @@
                         if (key == 290)
                         {
                             this.mc.gameSettings.hideGUI = !this.mc.gameSettings.hideGUI;
+                            this.mc.vrSettings.saveOptions();
                         }
                     }
 
@@ -536,7 +562,7 @@
         }
     }
 
-    private void onCharEvent(long windowPointer, int codePoint, int modifiers)
+    public void onCharEvent(long windowPointer, int codePoint, int modifiers)
     {
         if (windowPointer == this.mc.getMainWindow().getHandle())
         {
