--- a/net/minecraft/client/MainWindow.java
+++ b/net/minecraft/client/MainWindow.java
@@ -146,7 +146,7 @@
         GL.createCapabilities();
         this.updateVideoMode();
         this.updateFramebufferSize();
-        GLFW.glfwSetFramebufferSizeCallback(this.handle, this::onFramebufferSizeUpdate);
+   //     GLFW.glfwSetFramebufferSizeCallback(this.handle, this::onFramebufferSizeUpdate);
         GLFW.glfwSetWindowPosCallback(this.handle, this::onWindowPosUpdate);
         GLFW.glfwSetWindowSizeCallback(this.handle, this::onWindowSizeUpdate);
         GLFW.glfwSetWindowFocusCallback(this.handle, this::onWindowFocusUpdate);
@@ -308,7 +308,9 @@
     {
         RenderSystem.assertThread(RenderSystem::isOnRenderThreadOrInit);
         this.vsync = vsyncEnabled;
-        GLFW.glfwSwapInterval(vsyncEnabled ? 1 : 0);
+        // Viveraft - don't do vsync, silly billy
+        GLFW.glfwSwapInterval(0);
+        //
     }
 
     public void close()
@@ -327,21 +329,22 @@
         this.windowY = windowYIn;
     }
 
+    //TODO: remove all usages (Screenshots)
     private void onFramebufferSizeUpdate(long windowPointer, int framebufferWidth, int framebufferHeight)
     {
         if (windowPointer == this.handle)
         {
-            int i = this.getFramebufferWidth();
-            int j = this.getFramebufferHeight();
+            int i = this.getWidth();
+            int j = this.getHeight();
 
             if (framebufferWidth != 0 && framebufferHeight != 0)
             {
                 this.framebufferWidth = framebufferWidth;
                 this.framebufferHeight = framebufferHeight;
 
-                if (this.getFramebufferWidth() != i || this.getFramebufferHeight() != j)
+                if (this.getWidth() != i || this.getHeight() != j)
                 {
-                    this.mc.updateWindowSize();
+                	this.mc.updateWindowSize();
                 }
             }
         }
@@ -359,8 +362,13 @@
 
     private void onWindowSizeUpdate(long windowPointer, int windowWidthIn, int windowHeightIn)
     {
+    	//Vivecraft dont explode when minimized
+    	if(windowHeightIn*windowWidthIn == 0) return;
+    	//
+    	
         this.width = windowWidthIn;
         this.height = windowHeightIn;
+        this.mc.updateWindowSize();
     }
 
     private void onWindowFocusUpdate(long windowPointer, boolean hasFocus)
@@ -495,7 +503,7 @@
     {
         int i;
 
-        for (i = 1; i != guiScaleIn && i < this.framebufferWidth && i < this.framebufferHeight && this.framebufferWidth / (i + 1) >= 320 && this.framebufferHeight / (i + 1) >= 240; ++i)
+        for (i = 1; i != guiScaleIn && i < this.width && i < this.height && this.width / (i + 1) >= 320 && this.height / (i + 1) >= 240; ++i)
         {
         }
 
@@ -510,10 +518,10 @@
     public void setGuiScale(double scaleFactor)
     {
         this.guiScaleFactor = scaleFactor;
-        int i = (int)((double)this.framebufferWidth / scaleFactor);
-        this.scaledWidth = (double)this.framebufferWidth / scaleFactor > (double)i ? i + 1 : i;
-        int j = (int)((double)this.framebufferHeight / scaleFactor);
-        this.scaledHeight = (double)this.framebufferHeight / scaleFactor > (double)j ? j + 1 : j;
+        int i = (int)((double)this.width / scaleFactor);
+        this.scaledWidth = (double)this.width / scaleFactor > (double)i ? i + 1 : i;
+        int j = (int)((double)this.height / scaleFactor);
+        this.scaledHeight = (double)this.height / scaleFactor > (double)j ? j + 1 : j;
     }
 
     public void func_230148_b_(String p_230148_1_)
@@ -533,12 +541,16 @@
 
     public int getFramebufferWidth()
     {
-        return this.framebufferWidth;
+    	//Vivecraft experiment
+       	return Minecraft.getInstance().getFramebuffer().framebufferWidth;
+    	//
     }
 
     public int getFramebufferHeight()
     {
-        return this.framebufferHeight;
+    	//Vivecraft experiment
+       	return Minecraft.getInstance().getFramebuffer().framebufferHeight;
+    	//
     }
 
     public int getWidth()
