--- a/net/optifine/http/HttpUtils.java
+++ b/net/optifine/http/HttpUtils.java
@@ -9,6 +9,8 @@
 import java.net.HttpURLConnection;
 import java.net.URL;
 import java.util.Map;
+import java.util.Set;
+
 import net.minecraft.client.Minecraft;
 import net.optifine.Config;
 
@@ -88,7 +90,7 @@
 
             if (headers != null)
             {
-                for (String s : headers.keySet())
+                for (String s : (Set<String>)(Set<?>)headers.keySet())
                 {
                     String s1 = "" + headers.get(s);
                     httpurlconnection.setRequestProperty(s, s1);
