--- a/net/minecraft/client/renderer/entity/layers/MooshroomMushroomLayer.java
+++ b/net/minecraft/client/renderer/entity/layers/MooshroomMushroomLayer.java
@@ -31,7 +31,7 @@
     public MooshroomMushroomLayer(IEntityRenderer<T, CowModel<T>> rendererIn)
     {
         super(rendererIn);
-        MooshroomRenderer mooshroomrenderer = (MooshroomRenderer)rendererIn;
+        IEntityRenderer<T, CowModel<T>> mooshroomrenderer = rendererIn;
         this.modelRendererMushroom = new ModelRenderer(mooshroomrenderer.getEntityModel());
         this.modelRendererMushroom.setTextureSize(16, 16);
         this.modelRendererMushroom.rotationPointX = -6.0F;
