--- a/net/optifine/shaders/Shaders.java
+++ b/net/optifine/shaders/Shaders.java
@@ -124,13 +124,18 @@
 import org.lwjgl.opengl.EXTFramebufferObject;
 import org.lwjgl.opengl.GL;
 import org.lwjgl.opengl.GL11;
+import org.lwjgl.opengl.GL12;
+import org.lwjgl.opengl.GL14;
+import org.lwjgl.opengl.GL20;
+import org.lwjgl.opengl.GL30;
 import org.lwjgl.opengl.GL32;
 import org.lwjgl.opengl.GLCapabilities;
+import org.vivecraft.render.RenderPass;
 
 public class Shaders
 {
     static Minecraft mc;
-    static GameRenderer entityRenderer;
+    static GameRenderer gameRenderer;
     public static boolean isInitializedOnce = false;
     public static boolean isShaderPackInitialized = false;
     public static GLCapabilities capabilities;
@@ -339,12 +344,12 @@
     static final int MaxShadowColorBuffers = 8;
     static final int MaxShadowDepthBuffers = 2;
     static int usedColorBuffers = 0;
-    static int usedDepthBuffers = 0;
+    public static int usedDepthBuffers = 0;
     static int usedShadowColorBuffers = 0;
     static int usedShadowDepthBuffers = 0;
     static int usedColorAttachs = 0;
     static int usedDrawBuffers = 0;
-    static int dfb = 0;
+    public static int dfb = 0;
     static int sfb = 0;
     private static int[] gbuffersFormat = new int[8];
     public static boolean[] gbuffersClear = new boolean[8];
@@ -434,7 +439,7 @@
     private static ShaderOption[] shaderPackOptions = null;
     private static Set<String> shaderPackOptionSliders = null;
     static ShaderProfile[] shaderPackProfiles = null;
-    static Map<String, ScreenShaderOptions> shaderPackGuiScreens = null;
+    static Map<String, ScreenShaderOptions> shaderPackScreens = null;
     static Map<String, IExpressionBool> shaderPackProgramConditions = new HashMap<>();
     public static final String PATH_SHADERS_PROPERTIES = "/shaders/shaders.properties";
     public static PropertyDefaultFastFancyOff shaderPackClouds = new PropertyDefaultFastFancyOff("clouds", "Clouds", 0);
@@ -510,7 +515,7 @@
     static final FloatBuffer tempMatrixDirectBuffer = nextFloatBuffer(16);
     static final FloatBuffer tempDirectFloatBuffer = nextFloatBuffer(16);
     static final IntBuffer dfbColorTextures = nextIntBuffer(16);
-    static final IntBuffer dfbDepthTextures = nextIntBuffer(3);
+    public static final IntBuffer dfbDepthTextures = nextIntBuffer(3);
     static final IntBuffer sfbColorTextures = nextIntBuffer(8);
     static final IntBuffer sfbDepthTextures = nextIntBuffer(2);
     static final IntBuffer dfbDrawBuffers = nextIntBuffer(8);
@@ -529,6 +534,11 @@
     {
     }
 
+    //Vivecraft
+    static float[] sunPositionT = new float[4];
+    static float[] moonPositionT = new float[4];
+    //
+    
     private static ByteBuffer nextByteBuffer(int size)
     {
         ByteBuffer bytebuffer = bigBuffer;
@@ -831,7 +841,7 @@
             shaderPackOptions = null;
             shaderPackOptionSliders = null;
             shaderPackProfiles = null;
-            shaderPackGuiScreens = null;
+            shaderPackScreens = null;
             shaderPackProgramConditions.clear();
             shaderPackClouds.resetValue();
             shaderPackOldHandLight.resetValue();
@@ -887,6 +897,27 @@
         loadShaderPackResources();
         loadShaderPackDimensions();
         shaderPackOptions = loadShaderPackOptions();
+    
+        //Vivecraft clean shader options
+        for (ShaderOption sh : shaderPackOptions) {
+			System.out.println("Shader Option: " + sh.toString());
+			if (sh.getName().toLowerCase().contains("motion_blur")) //seus
+				sh.setValue("false");
+			if (sh.getName().toLowerCase().contains("motionblur")) //bsl, UShader
+				sh.setValue("false");
+			if (sh.getName().toLowerCase().contains("dof")) //bsl
+				sh.setValue("false");
+			if (sh.getName().toLowerCase().contains("dirtyLens")) //bsl
+				sh.setValue("false");
+			if (sh.getName().toLowerCase().contains("lensflare")) //bsl
+				sh.setValue("false");
+			if (sh.getName().toLowerCase().contains("idle_hands")) //qol
+				sh.setValue("false");
+			if (sh.getName().toLowerCase().contains("taa")) //sildurs
+				sh.setValue("false");
+		}
+        //
+        
         loadShaderPackProperties();
         boolean flag3 = shaderPackLoaded != flag;
         boolean flag4 = isOldLighting() != flag1;
@@ -1060,7 +1091,7 @@
                 shaderPackFrustumCulling.loadFrom(properties);
                 shaderPackOptionSliders = ShaderPackParser.parseOptionSliders(properties, shaderPackOptions);
                 shaderPackProfiles = ShaderPackParser.parseProfiles(properties, shaderPackOptions);
-                shaderPackGuiScreens = ShaderPackParser.parseGuiScreens(properties, shaderPackProfiles, shaderPackOptions);
+                shaderPackScreens = ShaderPackParser.parseGuiScreens(properties, shaderPackProfiles, shaderPackOptions);
                 shaderPackProgramConditions = ShaderPackParser.parseProgramConditions(properties, shaderPackOptions);
                 customTexturesGbuffers = loadCustomTextures(properties, 0);
                 customTexturesComposite = loadCustomTextures(properties, 1);
@@ -1562,7 +1593,7 @@
     {
         ShaderOption[] ashaderoption = (ShaderOption[])shaderPackOptions.clone();
 
-        if (shaderPackGuiScreens == null)
+        if (shaderPackScreens == null)
         {
             if (shaderPackProfiles != null)
             {
@@ -1575,7 +1606,7 @@
         else
         {
             String s = screenName != null ? "screen." + screenName : "screen";
-            ScreenShaderOptions screenshaderoptions = shaderPackGuiScreens.get(s);
+            ScreenShaderOptions screenshaderoptions = shaderPackScreens.get(s);
 
             if (screenshaderoptions == null)
             {
@@ -1596,7 +1627,7 @@
                     }
                     else if (shaderoption instanceof ShaderOptionRest)
                     {
-                        ShaderOption[] ashaderoption2 = getShaderOptionsRest(shaderPackGuiScreens, ashaderoption);
+                        ShaderOption[] ashaderoption2 = getShaderOptionsRest(shaderPackScreens, ashaderoption);
                         list.addAll(Arrays.asList(ashaderoption2));
                     }
                     else
@@ -1614,13 +1645,13 @@
     {
         String s = screenName != null ? "screen." + screenName : "screen";
 
-        if (shaderPackGuiScreens == null)
+        if (shaderPackScreens == null)
         {
             return def;
         }
         else
         {
-            ScreenShaderOptions screenshaderoptions = shaderPackGuiScreens.get(s);
+            ScreenShaderOptions screenshaderoptions = shaderPackScreens.get(s);
             return screenshaderoptions == null ? def : screenshaderoptions.getColumns();
         }
     }
@@ -2393,6 +2424,8 @@
             }
 
             checkGLError("Shaders.init");
+
+            mc.stereoProvider.reinitFrameBuffers("Shaders Initialized");
         }
     }
 
@@ -3510,8 +3543,14 @@
                 setProgramUniform1f(uniform_viewWidth, (float)renderWidth);
                 setProgramUniform1f(uniform_viewHeight, (float)renderHeight);
                 setProgramUniform1f(uniform_near, 0.05F);
-                setProgramUniform1f(uniform_far, (float)(mc.gameSettings.renderDistanceChunks * 16));
-                setProgramUniform3f(uniform_sunPosition, sunPosition[0], sunPosition[1], sunPosition[2]);
+                setProgramUniform1f(uniform_far, (float)(mc.gameSettings.renderDistanceChunks * 16));             
+                if(Minecraft.getInstance().currentPass == RenderPass.THIRD) {
+                    setProgramUniform3f(uniform_sunPosition, sunPositionT[0], sunPositionT[1], sunPositionT[2]);
+                    setProgramUniform3f(uniform_moonPosition, moonPositionT[0], moonPositionT[1], moonPositionT[2]);
+                }else {
+                    setProgramUniform3f(uniform_sunPosition, sunPosition[0], sunPosition[1], sunPosition[2]);
+                    setProgramUniform3f(uniform_moonPosition, moonPosition[0], moonPosition[1], moonPosition[2]);
+                }
                 setProgramUniform3f(uniform_moonPosition, moonPosition[0], moonPosition[1], moonPosition[2]);
                 setProgramUniform3f(uniform_shadowLightPosition, shadowLightPosition[0], shadowLightPosition[1], shadowLightPosition[2]);
                 setProgramUniform3f(uniform_upPosition, upPosition[0], upPosition[1], upPosition[2]);
@@ -3879,8 +3918,9 @@
     {
         renderDisplayWidth = mc.getMainWindow().getFramebufferWidth();
         renderDisplayHeight = mc.getMainWindow().getFramebufferHeight();
-        renderWidth = Math.round((float)renderDisplayWidth * configRenderResMul);
-        renderHeight = Math.round((float)renderDisplayHeight * configRenderResMul);
+        renderWidth = Math.round((float)renderDisplayWidth);
+        renderHeight = Math.round((float)renderDisplayHeight);
+        System.out.println("Shaders Render Res: " + renderWidth + " x " + renderHeight);
         setupFrameBuffer();
     }
 
@@ -3910,20 +3950,23 @@
         GL32.glDrawBuffers(0);
         GL32.glReadBuffer(0);
 
+        // VIVECRAFT: We need stencil buffers here
         for (int i = 0; i < usedDepthBuffers; ++i)
         {
             GlStateManager.bindTexture(dfbDepthTextures.get(i));
-            GL32.glTexParameteri(3553, 10242, 33071);
-            GL32.glTexParameteri(3553, 10243, 33071);
-            GL32.glTexParameteri(3553, 10241, 9728);
-            GL32.glTexParameteri(3553, 10240, 9728);
-            GL32.glTexParameteri(3553, 34891, 6409);
-            GL32.glTexImage2D(3553, 0, 6402, renderWidth, renderHeight, 0, 6402, 5126, (FloatBuffer)null);
+            GL11.glTexParameteri(GL11.GL_TEXTURE_2D, GL11.GL_TEXTURE_WRAP_S, GL12.GL_CLAMP_TO_EDGE);
+            GL11.glTexParameteri(GL11.GL_TEXTURE_2D, GL11.GL_TEXTURE_WRAP_T, GL12.GL_CLAMP_TO_EDGE);
+            GL11.glTexParameteri(GL11.GL_TEXTURE_2D, GL11.GL_TEXTURE_MIN_FILTER, GL11.GL_NEAREST);
+            GL11.glTexParameteri(GL11.GL_TEXTURE_2D, GL11.GL_TEXTURE_MAG_FILTER, GL11.GL_NEAREST);
+            GL11.glTexParameteri(GL11.GL_TEXTURE_2D, GL14.GL_DEPTH_TEXTURE_MODE, GL11.GL_LUMINANCE);
+            //GL11.glTexImage2D(GL11.GL_TEXTURE_2D, 0, GL11.GL_DEPTH_COMPONENT, renderWidth, renderHeight, 0, GL11.GL_DEPTH_COMPONENT, GL11.GL_FLOAT, (FloatBuffer)null);
+            GL11.glTexImage2D(GL11.GL_TEXTURE_2D, 0, GL30.GL_DEPTH24_STENCIL8, renderWidth, renderHeight, 0, GL30.GL_DEPTH_STENCIL, GL30.GL_UNSIGNED_INT_24_8, (FloatBuffer)null);
         }
 
-        EXTFramebufferObject.glFramebufferTexture2DEXT(36160, 36096, 3553, dfbDepthTextures.get(0), 0);
-        GL32.glDrawBuffers(dfbDrawBuffers);
-        GL32.glReadBuffer(0);
+        //EXTFramebufferObject.glFramebufferTexture2DEXT(36160, 36096, 3553, dfbDepthTextures.get(0), 0);
+        EXTFramebufferObject.glFramebufferTexture2DEXT(36160, GL30.GL_DEPTH_STENCIL_ATTACHMENT, 3553, dfbDepthTextures.get(0), 0);
+        GL20.glDrawBuffers(dfbDrawBuffers);
+        GL11.glReadBuffer(0);
         checkGLError("FT d");
 
         for (int k = 0; k < usedColorBuffers; ++k)
@@ -4081,7 +4124,7 @@
         checkWorldChanged(mc.world);
         mc = minecraft;
         mc.getProfiler().startSection("init");
-        entityRenderer = mc.gameRenderer;
+        gameRenderer = mc.gameRenderer;
 
         if (!isShaderPackInitialized)
         {
@@ -4101,9 +4144,10 @@
             }
         }
 
-        if (mc.getMainWindow().getFramebufferWidth() != renderDisplayWidth || mc.getMainWindow().getFramebufferHeight() != renderDisplayHeight)
+        if (mc.stereoProvider.reinitShadersFlag || mc.getMainWindow().getFramebufferWidth() != renderDisplayWidth || mc.getMainWindow().getFramebufferHeight() != renderDisplayHeight)
         {
             resize();
+            mc.stereoProvider.reinitShadersFlag  = false;
         }
 
         if (needResizeShadow)
@@ -4155,7 +4199,7 @@
 
             if (entity != null)
             {
-                isSleeping = entity instanceof LivingEntity && ((LivingEntity)entity).isSleeping();
+                isSleeping = false;//entity instanceof LivingEntity && ((LivingEntity)entity).isSleeping();
                 eyePosY = (float)activeRenderInfo.getProjectedView().getY();
                 eyeBrightness = mc.getRenderManager().getPackedLight(entity, partialTicks);
                 float f2 = (float)diffSystemTime * 0.01F;
@@ -4184,7 +4228,6 @@
 
                     if (livingentity.isPotionActive(Effects.NIGHT_VISION))
                     {
-                        GameRenderer gamerenderer = entityRenderer;
                         nightVision = GameRenderer.getNightVisionBrightness(livingentity, partialTicks);
                     }
 
@@ -4227,7 +4270,9 @@
         ((Buffer)previousModelView).position(0);
         ((Buffer)modelView).position(0);
         checkGLError("beginRender");
-        ShadersRender.renderShadowMap(entityRenderer, activeRenderInfo, 0, partialTicks, finishTimeNano);
+		// VIVECRAFT: Render shadow map once per frame instead of every damn pass
+		if (mc.currentPass == RenderPass.LEFT || mc.currentPass == RenderPass.THIRD)
+			ShadersRender.renderShadowMap(gameRenderer, gameRenderer.getActiveRenderInfo(), 0, partialTicks, finishTimeNano);
         mc.getProfiler().endSection();
         EXTFramebufferObject.glBindFramebufferEXT(36160, dfb);
 
@@ -4494,9 +4539,14 @@
         ((Buffer)projection).position(0);
         ((Buffer)projectionInverse).position(0);
         Matrix4f matrix4f = matrixStackIn.getLast().getMatrix();
-        Matrix4f matrix4f1 = new Matrix4f(matrix4f);
-        matrix4f1.transpose();
-        matrix4f1.write(tempMat);
+        //Vivecraft - I do not know why this is necessary to fix shadows. It should be in stereo already.
+        MatrixStack mat = new MatrixStack();
+        mat.getLast().getMatrix().setIdentity();
+        mat.getLast().getMatrix().mul(matrix4f);
+        gameRenderer.applystereo(mc.currentPass, mat);
+        //
+        mat.getLast().getMatrix().transpose();
+        mat.getLast().getMatrix().write(tempMat);
         ((Buffer)modelView).position(0);
         modelView.put(tempMat);
         SMath.invertMat4FBFA((FloatBuffer)((Buffer)modelViewInverse).position(0), (FloatBuffer)((Buffer)modelView).position(0), faModelViewInverse, faModelView);
@@ -4636,13 +4686,24 @@
         Matrix4f matrix4f1 = new Matrix4f(matrix4f);
         matrix4f1.transpose();
         matrix4f1.write(tempMat);
+    	if(Minecraft.getInstance().currentPass == RenderPass.THIRD) {
+    		//Vivecraft
+    		SMath.multiplyMat4xVec4(sunPositionT, tempMat, sunPosModelView);
+    		SMath.multiplyMat4xVec4(moonPositionT, tempMat, moonPosModelView);
+    		System.arraycopy(shadowAngle == sunAngle ? sunPositionT : moonPosition, 0, shadowLightPosition, 0, 3);
+    		setProgramUniform3f(uniform_sunPosition, sunPositionT[0], sunPositionT[1], sunPositionT[2]);
+    		setProgramUniform3f(uniform_moonPosition, moonPositionT[0], moonPositionT[1], moonPositionT[2]);
+    		setProgramUniform3f(uniform_shadowLightPosition, shadowLightPosition[0], shadowLightPosition[1], shadowLightPosition[2]);
+    		//
+    	}
+    	else {
         SMath.multiplyMat4xVec4(sunPosition, tempMat, sunPosModelView);
         SMath.multiplyMat4xVec4(moonPosition, tempMat, moonPosModelView);
         System.arraycopy(shadowAngle == sunAngle ? sunPosition : moonPosition, 0, shadowLightPosition, 0, 3);
         setProgramUniform3f(uniform_sunPosition, sunPosition[0], sunPosition[1], sunPosition[2]);
         setProgramUniform3f(uniform_moonPosition, moonPosition[0], moonPosition[1], moonPosition[2]);
         setProgramUniform3f(uniform_shadowLightPosition, shadowLightPosition[0], shadowLightPosition[1], shadowLightPosition[2]);
-
+	}
         if (customUniforms != null)
         {
             customUniforms.update();
@@ -5737,7 +5798,7 @@
     private static void loadShaderPackResources()
     {
         shaderPackResources = new HashMap<>();
-
+        System.out.println("loadShaderPackResources " + shaderPackLoaded);
         if (shaderPackLoaded)
         {
             List<String> list = new ArrayList<>();
@@ -5770,6 +5831,7 @@
                         {
                             String s6 = properties.getProperty(s5);
                             shaderPackResources.put(s5, s6);
+                            System.out.println("put " + s5 + " " + s6);
                         }
                     }
                 }
