--- a/net/optifine/config/ConnectedParser.java
+++ b/net/optifine/config/ConnectedParser.java
@@ -105,7 +105,7 @@
                 }
             }
 
-            return list.toArray(new MatchBlock[list.size()]);
+            return (MatchBlock[]) list.toArray(new MatchBlock[list.size()]);
         }
     }
 
@@ -376,7 +376,7 @@
 
     public static Comparable getPropertyValue(String value, Collection propertyValues)
     {
-        for (Comparable comparable : propertyValues)
+        for (Comparable comparable :(Set<Comparable>)(Set<?>) propertyValues)
         {
             if (getValueName(comparable).equals(value))
             {
@@ -491,7 +491,7 @@
                 list = list1;
             }
 
-            return list.toArray(new Biome[list.size()]);
+            return (Biome[]) list.toArray(new Biome[list.size()]);
         }
     }
 
