--- a/net/minecraft/client/tutorial/MovementStep.java
+++ b/net/minecraft/client/tutorial/MovementStep.java
@@ -93,12 +93,12 @@
             if (this.moveCompleted == -1 && this.moveToast == null)
             {
                 this.moveToast = new TutorialToast(TutorialToast.Icons.MOVEMENT_KEYS, MOVE_TITLE, MOVE_DESCRIPTION, true);
-                this.tutorial.getMinecraft().getToastGui().add(this.moveToast);
+                this.tutorial.getInstance().getToastGui().add(this.moveToast);
             }
             else if (this.moveCompleted != -1 && this.timeWaiting - this.moveCompleted >= 20 && this.lookCompleted == -1 && this.lookToast == null)
             {
                 this.lookToast = new TutorialToast(TutorialToast.Icons.MOUSE, LOOK_TITLE, LOOK_DESCRIPTION, true);
-                this.tutorial.getMinecraft().getToastGui().add(this.lookToast);
+                this.tutorial.getInstance().getToastGui().add(this.lookToast);
             }
         }
     }
