--- a/net/minecraft/client/tutorial/PunchTreeStep.java
+++ b/net/minecraft/client/tutorial/PunchTreeStep.java
@@ -38,7 +38,7 @@
         {
             if (this.timeWaiting == 1)
             {
-                ClientPlayerEntity clientplayerentity = this.tutorial.getMinecraft().player;
+                ClientPlayerEntity clientplayerentity = this.tutorial.getInstance().player;
 
                 if (clientplayerentity != null)
                 {
@@ -59,7 +59,7 @@
             if ((this.timeWaiting >= 600 || this.resetCount > 3) && this.toast == null)
             {
                 this.toast = new TutorialToast(TutorialToast.Icons.TREE, TITLE, DESCRIPTION, true);
-                this.tutorial.getMinecraft().getToastGui().add(this.toast);
+                this.tutorial.getInstance().getToastGui().add(this.toast);
             }
         }
     }
