--- a/net/minecraft/client/tutorial/Tutorial.java
+++ b/net/minecraft/client/tutorial/Tutorial.java
@@ -88,7 +88,9 @@
             this.stop();
         }
 
-        this.tutorialStep = this.minecraft.gameSettings.tutorialStep.create(this);
+        // VIVECRAFT: Real men don't need tutorials.
+        //this.tutorialStep = this.minecraft.gameSettings.tutorialStep.create(this);
+        this.tutorialStep = TutorialSteps.NONE.create(this);
     }
 
     public void tick()
@@ -122,7 +124,7 @@
         }
     }
 
-    public Minecraft getMinecraft()
+    public Minecraft getInstance()
     {
         return this.minecraft;
     }
