--- a/net/minecraft/client/tutorial/Tutorial.java
+++ b/net/minecraft/client/tutorial/Tutorial.java
@@ -101,7 +101,9 @@
             this.stop();
         }
 
-        this.tutorialStep = this.minecraft.gameSettings.tutorialStep.create(this);
+        // VIVECRAFT: Real men don't need tutorials.
+        //this.tutorialStep = this.minecraft.gameSettings.tutorialStep.create(this);
+        this.tutorialStep = TutorialSteps.NONE.create(this);
     }
 
     public void func_244698_a(TutorialToast p_244698_1_, int p_244698_2_)
@@ -158,7 +160,7 @@
         }
     }
 
-    public Minecraft getMinecraft()
+    public Minecraft getInstance()
     {
         return this.minecraft;
     }
