--- a/net/minecraft/client/main/Main.java
+++ b/net/minecraft/client/main/Main.java
@@ -38,12 +38,19 @@
 public class Main
 {
     private static final Logger LOGGER = LogManager.getLogger();
-
+	public static boolean kiosk;
+	public static boolean viewonly;
+	public static boolean katvr;
+	public static boolean infinadeck;
     public static void main(String[] p_main_0_)
     {
         OptionParser optionparser = new OptionParser();
         optionparser.allowsUnrecognizedOptions();
         optionparser.accepts("demo");
+        optionparser.accepts("kiosk");
+        optionparser.accepts("viewonly");
+        optionparser.accepts("katvr");
+        optionparser.accepts("infinadeck");
         optionparser.accepts("fullscreen");
         optionparser.accepts("checkGlErrors");
         OptionSpec<String> optionspec = optionparser.accepts("server").withRequiredArg();
@@ -59,8 +66,8 @@
         OptionSpec<String> optionspec10 = optionparser.accepts("uuid").withRequiredArg();
         OptionSpec<String> optionspec11 = optionparser.accepts("accessToken").withRequiredArg().required();
         OptionSpec<String> optionspec12 = optionparser.accepts("version").withRequiredArg().required();
-        OptionSpec<Integer> optionspec13 = optionparser.accepts("width").withRequiredArg().ofType(Integer.class).defaultsTo(854);
-        OptionSpec<Integer> optionspec14 = optionparser.accepts("height").withRequiredArg().ofType(Integer.class).defaultsTo(480);
+        OptionSpec<Integer> optionspec13 = optionparser.accepts("width").withRequiredArg().ofType(Integer.class).defaultsTo(1280);
+        OptionSpec<Integer> optionspec14 = optionparser.accepts("height").withRequiredArg().ofType(Integer.class).defaultsTo(720);
         OptionSpec<Integer> optionspec15 = optionparser.accepts("fullscreenWidth").withRequiredArg().ofType(Integer.class);
         OptionSpec<Integer> optionspec16 = optionparser.accepts("fullscreenHeight").withRequiredArg().ofType(Integer.class);
         OptionSpec<String> optionspec17 = optionparser.accepts("userProperties").withRequiredArg().defaultsTo("{}");
@@ -112,6 +119,17 @@
         OptionalInt optionalint1 = toOptionalInt(getValue(optionset, optionspec16));
         boolean flag = optionset.has("fullscreen");
         boolean flag1 = optionset.has("demo");
+        //Vivecraft
+        kiosk = optionset.has("kiosk");
+    	if(kiosk) System.out.println("Setting kiosk");
+        if (kiosk) {
+        	viewonly = optionset.has("viewonly");
+        	if(viewonly) 
+        		System.out.println("Setting viewonly");
+        }
+        katvr = optionset.has("katvr");
+        infinadeck = optionset.has("infinadeck");
+        //
         String s3 = getValue(optionset, optionspec12);
         Gson gson = (new GsonBuilder()).registerTypeAdapter(PropertyMap.class, new Serializer()).create();
         PropertyMap propertymap = JSONUtils.fromJson(gson, getValue(optionset, optionspec17), PropertyMap.class);
@@ -267,6 +285,7 @@
 
     static
     {
-        System.setProperty("java.awt.headless", "true");
+        // wtf mojang
+        //System.setProperty("java.awt.headless", "true");
     }
 }
