--- a/net/optifine/util/NativeMemory.java
+++ b/net/optifine/util/NativeMemory.java
@@ -75,7 +75,7 @@
             Class oclass = Class.forName(path[0]);
             Method method = oclass.getMethod(path[1]);
             method.setAccessible(true);
-            final Object object = null;
+            Object object = null;
 
             for (int i = 2; i < path.length; ++i)
             {
@@ -86,6 +86,7 @@
             }
 
             final Method method1 = method;
+            final Object object1 = object;
             return new LongSupplier()
             {
                 private boolean disabled = false;
@@ -99,7 +100,8 @@
                     {
                         try
                         {
-                            return method1.invoke(object);
+                    
+                            return (long) method1.invoke(object1);
                         }
                         catch (Throwable throwable)
                         {
