--- a/net/optifine/util/NativeMemory.java
+++ b/net/optifine/util/NativeMemory.java
@@ -76,7 +76,7 @@
             Class oclass = Class.forName(path[0]);
             Method method = oclass.getMethod(path[1]);
             method.setAccessible(true);
-            final Object object = null;
+            Object object = null;
 
             for (int i = 2; i < path.length; ++i)
             {
@@ -85,8 +85,10 @@
                 method = object.getClass().getMethod(s);
                 method.setAccessible(true);
             }
-
+            
+        	final Object o = object;
             final Method method1 = method;
+            
             LongSupplier longsupplier = new LongSupplier()
             {
                 private boolean disabled = false;
@@ -100,7 +102,8 @@
                     {
                         try
                         {
-                            return method1.invoke(object);
+                    
+                            return (long) method1.invoke(o);
                         }
                         catch (Throwable throwable)
                         {
