--- a/net/optifine/util/ResUtils.java
+++ b/net/optifine/util/ResUtils.java
@@ -150,7 +150,7 @@
                 }
             }
 
-            return list.toArray(new String[list.size()]);
+            return (String[]) list.toArray(new String[list.size()]);
         }
     }
 
@@ -204,7 +204,7 @@
                 }
             }
 
-            return list.toArray(new String[list.size()]);
+            return (String[]) list.toArray(new String[list.size()]);
         }
     }
 
@@ -242,7 +242,7 @@
             }
 
             zipfile.close();
-            String[] astring = list.toArray(new String[list.size()]);
+            String[] astring = (String[]) list.toArray(new String[list.size()]);
             return astring;
         }
         catch (IOException ioexception)
