--- a/net/minecraft/client/renderer/tileentity/ItemStackTileEntityRenderer.java
+++ b/net/minecraft/client/renderer/tileentity/ItemStackTileEntityRenderer.java
@@ -14,9 +14,11 @@
 import net.minecraft.block.Block;
 import net.minecraft.block.Blocks;
 import net.minecraft.block.ShulkerBoxBlock;
+import net.minecraft.client.Minecraft;
 import net.minecraft.client.renderer.IRenderTypeBuffer;
 import net.minecraft.client.renderer.ItemRenderer;
 import net.minecraft.client.renderer.LightTexture;
+import net.minecraft.client.renderer.RenderType;
 import net.minecraft.client.renderer.entity.model.ShieldModel;
 import net.minecraft.client.renderer.entity.model.TridentModel;
 import net.minecraft.client.renderer.model.ItemCameraTransforms;
@@ -43,6 +45,7 @@
 import net.minecraft.util.Direction;
 import net.optifine.EmissiveTextures;
 import org.apache.commons.lang3.StringUtils;
+import org.vivecraft.gameplay.trackers.SwingTracker;
 
 public class ItemStackTileEntityRenderer
 {
@@ -166,34 +169,54 @@
                 TileEntityRendererDispatcher.instance.renderItem(tileentity, p_renderRaw_2_, p_renderRaw_3_, p_renderRaw_4_, p_renderRaw_5_);
             }
         }
-        else if (item == Items.SHIELD)
+        else
         {
-            boolean flag = p_renderRaw_1_.getChildTag("BlockEntityTag") != null;
-            p_renderRaw_2_.push();
-            p_renderRaw_2_.scale(1.0F, -1.0F, -1.0F);
-            RenderMaterial rendermaterial = flag ? ModelBakery.LOCATION_SHIELD_BASE : ModelBakery.LOCATION_SHIELD_NO_PATTERN;
-            IVertexBuilder ivertexbuilder1 = rendermaterial.getSprite().wrapBuffer(ItemRenderer.func_239391_c_(p_renderRaw_3_, this.modelShield.getRenderType(rendermaterial.getAtlasLocation()), true, p_renderRaw_1_.hasEffect()));
-            this.modelShield.func_228294_b_().render(p_renderRaw_2_, ivertexbuilder1, p_renderRaw_4_, p_renderRaw_5_, 1.0F, 1.0F, 1.0F, 1.0F);
-
-            if (flag)
-            {
-                List<Pair<BannerPattern, DyeColor>> list = BannerTileEntity.func_230138_a_(ShieldItem.getColor(p_renderRaw_1_), BannerTileEntity.func_230139_a_(p_renderRaw_1_));
-                BannerTileEntityRenderer.func_241717_a_(p_renderRaw_2_, p_renderRaw_3_, p_renderRaw_4_, p_renderRaw_5_, this.modelShield.func_228293_a_(), rendermaterial, false, list, p_renderRaw_1_.hasEffect());
-            }
-            else
-            {
-                this.modelShield.func_228293_a_().render(p_renderRaw_2_, ivertexbuilder1, p_renderRaw_4_, p_renderRaw_5_, 1.0F, 1.0F, 1.0F, 1.0F);
-            }
-
-            p_renderRaw_2_.pop();
-        }
-        else if (item == Items.TRIDENT)
-        {
-            p_renderRaw_2_.push();
-            p_renderRaw_2_.scale(1.0F, -1.0F, -1.0F);
-            IVertexBuilder ivertexbuilder = ItemRenderer.func_239391_c_(p_renderRaw_3_, this.trident.getRenderType(TridentModel.TEXTURE_LOCATION), false, p_renderRaw_1_.hasEffect());
-            this.trident.render(p_renderRaw_2_, ivertexbuilder, p_renderRaw_4_, p_renderRaw_5_, 1.0F, 1.0F, 1.0F, 1.0F);
-            p_renderRaw_2_.pop();
+        	//Vivecraft
+        	float a = 1;
+        	if(Minecraft.getInstance().getItemRenderer().isfphand) {
+        		a = SwingTracker.getItemFade(Minecraft.getInstance().player, p_renderRaw_1_);
+        	}
+        	//
+
+        	if (item == Items.SHIELD)
+        	{
+        		boolean flag = p_renderRaw_1_.getChildTag("BlockEntityTag") != null;
+        		p_renderRaw_2_.push();
+        		p_renderRaw_2_.scale(1.0F, -1.0F, -1.0F);
+        		RenderMaterial rendermaterial = flag ? ModelBakery.LOCATION_SHIELD_BASE : ModelBakery.LOCATION_SHIELD_NO_PATTERN;
+        		IVertexBuilder ivertexbuilder1 = rendermaterial.getSprite().wrapBuffer(ItemRenderer.func_239391_c_(p_renderRaw_3_, this.modelShield.getRenderType(rendermaterial.getAtlasLocation()), true, p_renderRaw_1_.hasEffect()));
+        		//Vivecraft
+        		if(a<1)
+        			ivertexbuilder1 = rendermaterial.getSprite().wrapBuffer(ItemRenderer.func_239391_c_(p_renderRaw_3_, RenderType.getEntityTranslucent(rendermaterial.getAtlasLocation()), true, p_renderRaw_1_.hasEffect()));
+        		this.modelShield.func_228294_b_().render(p_renderRaw_2_, ivertexbuilder1, p_renderRaw_4_, p_renderRaw_5_, 1.0F, 1.0F, 1.0F, a);
+        		//
+        		
+        		if (flag)
+        		{
+        			List<Pair<BannerPattern, DyeColor>> list = BannerTileEntity.func_230138_a_(ShieldItem.getColor(p_renderRaw_1_), BannerTileEntity.func_230139_a_(p_renderRaw_1_));
+        			BannerTileEntityRenderer.func_241717_a_(p_renderRaw_2_, p_renderRaw_3_, p_renderRaw_4_, p_renderRaw_5_, this.modelShield.func_228293_a_(), rendermaterial, false, list, p_renderRaw_1_.hasEffect());
+        		}
+        		else
+        		{
+        			//Vivecraft a
+        			this.modelShield.func_228293_a_().render(p_renderRaw_2_, ivertexbuilder1, p_renderRaw_4_, p_renderRaw_5_, 1.0F, 1.0F, 1.0F, a);
+        			//
+        		}
+
+        		p_renderRaw_2_.pop();
+        	}
+        	else if (item == Items.TRIDENT)
+        	{
+        		p_renderRaw_2_.push();
+        		p_renderRaw_2_.scale(1.0F, -1.0F, -1.0F);
+        		IVertexBuilder ivertexbuilder = ItemRenderer.func_239391_c_(p_renderRaw_3_, this.trident.getRenderType(TridentModel.TEXTURE_LOCATION), false, p_renderRaw_1_.hasEffect());
+        		//Vivecraft
+        		if(a<1)
+        			ivertexbuilder = ItemRenderer.func_239391_c_(p_renderRaw_3_, RenderType.getEntityTranslucent(TridentModel.TEXTURE_LOCATION), false, p_renderRaw_1_.hasEffect());
+        		this.trident.render(p_renderRaw_2_, ivertexbuilder, p_renderRaw_4_, p_renderRaw_5_, 1.0F, 1.0F, 1.0F, a);
+        		//
+        		p_renderRaw_2_.pop();
+        	}
         }
     }
 }
