--- a/net/optifine/shaders/Shaders.java
+++ b/net/optifine/shaders/Shaders.java
@@ -118,6 +118,7 @@
 import net.optifine.util.WorldUtils;
 import org.apache.commons.io.IOUtils;
 import org.lwjgl.BufferUtils;
+import org.lwjgl.opengl.ARBFragmentShader;
 import org.lwjgl.opengl.ARBGeometryShader4;
 import org.lwjgl.opengl.ARBShaderObjects;
 import org.lwjgl.opengl.ARBVertexShader;
@@ -811,6 +812,7 @@
 
     public static void loadShaderPack()
     {
+    	mc = Minecraft.getInstance();
         boolean flag = shaderPackLoaded;
         boolean flag1 = isOldLighting();
 
@@ -1090,7 +1092,7 @@
         Set set = props.keySet();
         List<ICustomTexture> list = new ArrayList<>();
 
-        for (String s1 : set)
+        for (String s1 : (Set<String>)(Set<?>)set)
         {
             if (s1.startsWith(s))
             {
@@ -2453,11 +2455,11 @@
 
     private static int getDrawBuffer(Program p, String str, int i)
     {
-        int i = 0;
+        int d = 0;
 
         if (i >= str.length())
         {
-            return i;
+            return d;
         }
         else
         {
@@ -4304,15 +4306,15 @@
         GlStateManager.activeTexture(33984);
     }
 
-    public static void checkWorldChanged(ClientWorld world)
+    public static void checkWorldChanged(ClientWorld worldin)
     {
-        if (currentWorld != world)
+        if (currentWorld != worldin)
         {
-            World world = currentWorld;
-            currentWorld = world;
+            World worldold = currentWorld;
+            currentWorld = worldin;
             setCameraOffset(mc.getRenderViewEntity());
-            int i = WorldUtils.getDimensionId(world);
-            int j = WorldUtils.getDimensionId(world);
+            int i = WorldUtils.getDimensionId(worldold);
+            int j = WorldUtils.getDimensionId(worldin);
 
             if (j != i)
             {
@@ -5766,7 +5768,7 @@
                         Lang.loadLocaleData(inputstream, properties);
                         inputstream.close();
 
-                        for (String s5 : properties.keySet())
+                        for (String s5 : (Set<String>)(Set<?>)properties.keySet())
                         {
                             String s6 = properties.getProperty(s5);
                             shaderPackResources.put(s5, s6);
