--- a/net/optifine/shaders/Shaders.java
+++ b/net/optifine/shaders/Shaders.java
@@ -124,6 +124,7 @@
 import net.optifine.util.WorldUtils;
 import org.apache.commons.io.IOUtils;
 import org.lwjgl.BufferUtils;
+import org.lwjgl.opengl.ARBFragmentShader;
 import org.lwjgl.opengl.ARBGeometryShader4;
 import org.lwjgl.opengl.ARBShaderObjects;
 import org.lwjgl.opengl.ARBVertexShader;
@@ -871,6 +872,7 @@
 
     public static void loadShaderPack()
     {
+    	mc = Minecraft.getInstance();
         boolean flag = shaderPackLoaded;
         boolean flag1 = isOldLighting();
 
@@ -1184,7 +1186,7 @@
         Set set = props.keySet();
         List<ICustomTexture> list = new ArrayList<>();
 
-        for (String s1 : set)
+        for (String s1 : (Set<String>)(Set<?>)set)
         {
             if (s1.startsWith(s))
             {
@@ -4834,12 +4836,12 @@
         bindCustomTextures(customTextures);
     }
 
-    public static void checkWorldChanged(ClientWorld world)
+    public static void checkWorldChanged(ClientWorld worldin)
     {
-        if (currentWorld != world)
+        if (currentWorld != worldin)
         {
             World world = currentWorld;
-            currentWorld = world;
+            currentWorld = worldin;
 
             if (currentWorld == null)
             {
@@ -4853,7 +4855,7 @@
 
             setCameraOffset(mc.getRenderViewEntity());
             int i = WorldUtils.getDimensionId(world);
-            int j = WorldUtils.getDimensionId(world);
+            int j = WorldUtils.getDimensionId(worldin);
 
             if (j != i)
             {
@@ -6323,7 +6325,7 @@
                         Lang.loadLocaleData(inputstream, properties);
                         inputstream.close();
 
-                        for (String s5 : properties.keySet())
+                        for (String s5 : (Set<String>)(Set<?>)properties.keySet())
                         {
                             String s6 = properties.getProperty(s5);
                             shaderPackResources.put(s5, s6);
