--- a/net/optifine/shaders/Shaders.java
+++ b/net/optifine/shaders/Shaders.java
@@ -114,6 +114,7 @@
 import net.optifine.util.TimedEvent;
 import org.apache.commons.io.IOUtils;
 import org.lwjgl.BufferUtils;
+import org.lwjgl.opengl.ARBFragmentShader;
 import org.lwjgl.opengl.ARBGeometryShader4;
 import org.lwjgl.opengl.ARBShaderObjects;
 import org.lwjgl.opengl.ARBVertexShader;
@@ -809,6 +810,7 @@
 
     public static void loadShaderPack()
     {
+    	mc = Minecraft.getInstance();
         boolean flag = shaderPackLoaded;
         boolean flag1 = isOldLighting();
 
@@ -1088,7 +1090,7 @@
         Set set = props.keySet();
         List<ICustomTexture> list = new ArrayList<>();
 
-        for (String s1 : set)
+        for (String s1 : (Set<String>)(Set<?>)set)
         {
             if (s1.startsWith(s))
             {
@@ -2047,10 +2049,10 @@
         SMCLog.info(stringbuilder.toString());
     }
 
-    public static void startup(Minecraft mc)
+    public static void startup(Minecraft mcin)
     {
         checkShadersModInstalled();
-        mc = mc;
+        mc = mcin;
         mc = Minecraft.getInstance();
         capabilities = GL.getCapabilities();
         glVersionString = GL11.glGetString(GL11.GL_VERSION);
@@ -2462,11 +2464,11 @@
 
     private static int getDrawBuffer(Program p, String str, int i)
     {
-        int i = 0;
+        int d = 0;
 
         if (i >= str.length())
         {
-            return i;
+            return d;
         }
         else
         {
@@ -4314,14 +4316,14 @@
         GlStateManager.activeTexture(33984);
     }
 
-    public static void checkWorldChanged(ClientWorld world)
+    public static void checkWorldChanged(ClientWorld worldin)
     {
-        if (currentWorld != world)
+        if (currentWorld != worldin)
         {
             World world = currentWorld;
-            currentWorld = world;
+            currentWorld = worldin;
             setCameraOffset(mc.getRenderViewEntity());
-            int i = getDimensionId(world);
+            int i = getDimensionId(worldin);
             int j = getDimensionId(world);
 
             if (j != i)
@@ -5765,7 +5767,7 @@
                         Lang.loadLocaleData(inputstream, properties);
                         inputstream.close();
 
-                        for (String s5 : properties.keySet())
+                        for (String s5 : (Set<String>)(Set<?>)properties.keySet())
                         {
                             String s6 = properties.getProperty(s5);
                             shaderPackResources.put(s5, s6);
