--- a/net/minecraft/client/resources/ClientLanguageMap.java
+++ b/net/minecraft/client/resources/ClientLanguageMap.java
@@ -17,6 +17,7 @@
 import net.optifine.Lang;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
+import org.vivecraft.utils.LangHelper;
 
 public class ClientLanguageMap extends LanguageMap
 {
@@ -47,6 +48,7 @@
                     ResourceLocation resourcelocation = new ResourceLocation(s1, s);
                     func_239498_a_(p_239497_0_.getAllResources(resourcelocation), map);
                     Lang.loadResources(p_239497_0_, language.getCode(), map);
+                    LangHelper.loadLocaleData(language.getCode(), map); // VIVECRAFT
                 }
                 catch (FileNotFoundException filenotfoundexception)
                 {
