--- a/net/optifine/Config.java
+++ b/net/optifine/Config.java
@@ -1185,7 +1185,7 @@
             }
         }
 
-        IResourcePack[] airesourcepack = list.toArray(new IResourcePack[list.size()]);
+        IResourcePack[] airesourcepack = (IResourcePack[]) list.toArray(new IResourcePack[list.size()]);
         return airesourcepack;
     }
 
@@ -1510,7 +1510,7 @@
             list.add(s);
         }
 
-        String[] astring = list.toArray(new String[list.size()]);
+        String[] astring = (String[]) list.toArray(new String[list.size()]);
         return astring;
     }
 
@@ -1642,7 +1642,7 @@
 
             if (s == null)
             {
-                return list.toArray(new String[list.size()]);
+                return (String[]) list.toArray(new String[list.size()]);
             }
 
             list.add(s);
@@ -1938,7 +1938,7 @@
         {
             int i = arr.length;
             int j = i + 1;
-            Object[] aobject = Array.newInstance(arr.getClass().getComponentType(), j);
+            Object[] aobject = (Object[]) Array.newInstance(arr.getClass().getComponentType(), j);
             System.arraycopy(arr, 0, aobject, 0, i);
             aobject[i] = obj;
             return aobject;
@@ -1949,7 +1949,7 @@
     {
         List list = new ArrayList<>(Arrays.asList(arr));
         list.add(index, obj);
-        Object[] aobject = Array.newInstance(arr.getClass().getComponentType(), list.size());
+        Object[] aobject = (Object[]) Array.newInstance(arr.getClass().getComponentType(), list.size());
         return list.toArray(aobject);
     }
 
@@ -1967,7 +1967,7 @@
         {
             int i = arr.length;
             int j = i + objs.length;
-            Object[] aobject = Array.newInstance(arr.getClass().getComponentType(), j);
+            Object[] aobject = (Object[]) Array.newInstance(arr.getClass().getComponentType(), j);
             System.arraycopy(arr, 0, aobject, 0, i);
             System.arraycopy(objs, 0, aobject, i, objs.length);
             return aobject;
@@ -1997,7 +1997,7 @@
         }
         else
         {
-            Object[] aobject = Array.newInstance(elementClass, coll.size());
+            Object[] aobject = (Object[]) Array.newInstance(elementClass, coll.size());
             return coll.toArray(aobject);
         }
     }
