--- a/net/optifine/ConnectedProperties.java
+++ b/net/optifine/ConnectedProperties.java
@@ -6,6 +6,7 @@
 import java.util.List;
 import java.util.Map;
 import java.util.Properties;
+import java.util.Set;
 import net.minecraft.block.Block;
 import net.minecraft.block.BlockState;
 import net.minecraft.block.Blocks;
@@ -312,7 +313,7 @@
                 list.add(s);
             }
 
-            String[] astring2 = list.toArray(new String[list.size()]);
+            String[] astring2 = (String[]) list.toArray(new String[list.size()]);
 
             for (int i1 = 0; i1 < astring2.length; ++i1)
             {
@@ -474,7 +475,7 @@
 
     public static Property getProperty(String key, Collection properties)
     {
-        for (Property property : properties)
+        for (Property property : (Iterable<Property>)(Iterable<?>)properties)
         {
             if (key.equals(property.getName()))
             {
@@ -1227,7 +1228,7 @@
                 }
             }
 
-            return list.toArray(new TextureAtlasSprite[list.size()]);
+            return (TextureAtlasSprite[]) list.toArray(new TextureAtlasSprite[list.size()]);
         }
     }
 
