--- a/net/optifine/CustomGuiProperties.java
+++ b/net/optifine/CustomGuiProperties.java
@@ -2,6 +2,7 @@
 
 import java.util.HashMap;
 import java.util.Map;
+import java.util.Set;
 import java.util.Properties;
 import net.minecraft.block.BlockState;
 import net.minecraft.block.ChestBlock;
@@ -205,7 +206,7 @@
 
         String s5 = property + ".";
 
-        for (String s1 : props.keySet())
+        for (String s1 : (Set<String>)(Set<?>)props.keySet())
         {
             if (s1.startsWith(s5))
             {
