--- a/net/optifine/CustomColors.java
+++ b/net/optifine/CustomColors.java
@@ -469,7 +469,7 @@
         String s = "palette.block.";
         Map map = new HashMap();
 
-        for (String s1 : props.keySet())
+        for (String s1 : (Set<String>)(Set<?>)props.keySet())
         {
             String s2 = props.getProperty(s1);
 
@@ -479,7 +479,7 @@
             }
         }
 
-        String[] astring = map.keySet().toArray(new String[map.size()]);
+        String[] astring = (String[]) map.keySet().toArray(new String[map.size()]);
 
         for (int j = 0; j < astring.length; ++j)
         {
@@ -517,7 +517,7 @@
             }
         }
 
-        return list.size() <= 0 ? null : list.toArray(new CustomColormap[list.size()]);
+        return list.size() <= 0 ? null : (CustomColormap[])list.toArray(new CustomColormap[list.size()]);
     }
 
     private static CustomColormap[][] readBlockColormaps(String[] basePaths, CustomColormap[] basePalettes, int width, int height)
@@ -607,15 +607,15 @@
             list.add((Object)null);
         }
 
-        List list = (List)list.get(id);
+        List sublist = (List)list.get(id);
 
-        if (list == null)
+        if (sublist == null)
         {
-            list = new ArrayList();
-            list.set(id, list);
+            sublist = new ArrayList();
+            list.set(id, sublist);
         }
 
-        list.add(cm);
+        sublist.add(cm);
     }
 
     private static CustomColormap[][] blockListToArray(List list)
@@ -624,11 +624,11 @@
 
         for (int i = 0; i < list.size(); ++i)
         {
-            List list = (List)list.get(i);
+            List lista = (List)list.get(i);
 
-            if (list != null)
+            if (lista != null)
             {
-                CustomColormap[] acustomcolormap1 = list.toArray(new CustomColormap[list.size()]);
+                CustomColormap[] acustomcolormap1 = (CustomColormap[]) lista.toArray(new CustomColormap[lista.size()]);
                 acustomcolormap[i] = acustomcolormap1;
             }
         }
@@ -1353,7 +1353,7 @@
         Set set = props.keySet();
         int i = 0;
 
-        for (String s : set)
+        for (String s : (Set<String>)(Set<?>)set)
         {
             String s1 = props.getProperty(s);
 
@@ -1499,7 +1499,7 @@
         float[][] afloat1 = new float[adyecolor.length][];
         int k = 0;
 
-        for (String s : props.keySet())
+        for (String s : (Set<String>)(Set<?>)props.keySet())
         {
             String s1 = props.getProperty(s);
 
@@ -1566,7 +1566,7 @@
         Arrays.fill(aint, -1);
         int i = 0;
 
-        for (String s : props.keySet())
+        for (String s : (Set<String>)(Set<?>)props.keySet())
         {
             String s1 = props.getProperty(s);
 
@@ -1622,7 +1622,7 @@
         Arrays.fill(aint, -1);
         int i = 0;
 
-        for (String s : props.keySet())
+        for (String s : (Set<String>)(Set<?>)props.keySet())
         {
             String s1 = props.getProperty(s);
 
@@ -1661,7 +1661,7 @@
         Arrays.fill(aint, -1);
         int i = 0;
 
-        for (String s : props.keySet())
+        for (String s : (Set<String>)(Set<?>)props.keySet())
         {
             String s1 = props.getProperty(s);
 
