--- a/net/optifine/TextureAnimations.java
+++ b/net/optifine/TextureAnimations.java
@@ -91,7 +91,7 @@
             }
         }
 
-        TextureAnimation[] atextureanimation1 = list.toArray(new TextureAnimation[list.size()]);
+        TextureAnimation[] atextureanimation1 = (TextureAnimation[]) list.toArray(new TextureAnimation[list.size()]);
         return atextureanimation1;
     }
 
@@ -144,7 +144,7 @@
                 }
             }
 
-            TextureAnimation[] atextureanimation = list.toArray(new TextureAnimation[list.size()]);
+            TextureAnimation[] atextureanimation = (TextureAnimation[]) list.toArray(new TextureAnimation[list.size()]);
             return atextureanimation;
         }
     }
