--- a/net/optifine/override/PlayerControllerOF.java
+++ b/net/optifine/override/PlayerControllerOF.java
@@ -52,7 +52,7 @@
         return actionresulttype;
     }
 
-    public ActionResultType func_217292_a(ClientPlayerEntity player, ClientWorld worldIn, Hand hand, BlockRayTraceResult rayTrace)
+    public ActionResultType processRightClickBlock(ClientPlayerEntity player, ClientWorld worldIn, Hand hand, BlockRayTraceResult rayTrace)
     {
         this.acting = true;
         this.lastClickBlockPos = rayTrace.getPos();
