--- a/net/optifine/override/PlayerControllerOF.java
+++ b/net/optifine/override/PlayerControllerOF.java
@@ -1,5 +1,7 @@
 package net.optifine.override;
 
+import org.vivecraft.api.NetworkHelper;
+
 import net.minecraft.client.Minecraft;
 import net.minecraft.client.entity.player.ClientPlayerEntity;
 import net.minecraft.client.multiplayer.PlayerController;
@@ -7,6 +9,7 @@
 import net.minecraft.client.world.ClientWorld;
 import net.minecraft.entity.Entity;
 import net.minecraft.entity.player.PlayerEntity;
+import net.minecraft.network.play.client.CPlayerDiggingPacket;
 import net.minecraft.util.ActionResultType;
 import net.minecraft.util.Direction;
 import net.minecraft.util.Hand;
@@ -29,6 +32,7 @@
     /**
      * Called when the player is hitting a block with an item.
      */
+    @Override
     public boolean clickBlock(BlockPos loc, Direction face)
     {
         this.acting = true;
@@ -37,7 +41,8 @@
         this.acting = false;
         return flag;
     }
-
+    
+    @Override
     public boolean onPlayerDamageBlock(BlockPos posBlock, Direction directionFacing)
     {
         this.acting = true;
@@ -47,16 +52,33 @@
         return flag;
     }
 
+    @Override
     public ActionResultType processRightClick(PlayerEntity player, World worldIn, Hand hand)
     {
         this.acting = true;
+		//Vivecraft
+		NetworkHelper.overrideLook(player, Minecraft.getInstance().vrPlayer.getRightClickLookOverride(player, hand.ordinal()));
+		//	
         ActionResultType actionresulttype = super.processRightClick(player, worldIn, hand);
         this.acting = false;
         return actionresulttype;
     }
 
+    //Vivecraft
+    @Override
+    public void onStoppedUsingItem(PlayerEntity playerIn)
+    {
+		NetworkHelper.overrideLook(playerIn, Minecraft.getInstance().vrPlayer.getRightClickLookOverride(playerIn, playerIn.getActiveHand().ordinal()));
+    	super.onStoppedUsingItem(playerIn);
+    }
+    //
+
+    @Override
     public ActionResultType func_217292_a(ClientPlayerEntity player, ClientWorld worldIn, Hand hand, BlockRayTraceResult rayTrace)
     {
+		//Vivecraft
+		NetworkHelper.overrideLook(player, rayTrace.getHitVec().subtract(player.getEyePosition(1)).normalize());
+		//	
         this.acting = true;
         this.lastClickBlockPos = rayTrace.getPos();
         ActionResultType actionresulttype = super.func_217292_a(player, worldIn, hand, rayTrace);
@@ -67,6 +89,7 @@
     /**
      * Handles right clicking an entity, sends a packet to the server.
      */
+    @Override
     public ActionResultType interactWithEntity(PlayerEntity player, Entity target, Hand hand)
     {
         this.lastClickEntity = target;
@@ -76,6 +99,7 @@
     /**
      * Handles right clicking an entity from the entities side, sends a packet to the server.
      */
+    @Override
     public ActionResultType interactWithEntity(PlayerEntity player, Entity target, EntityRayTraceResult ray, Hand hand)
     {
         this.lastClickEntity = target;
