--- a/net/minecraft/client/network/play/ClientPlayNetHandler.java
+++ b/net/minecraft/client/network/play/ClientPlayNetHandler.java
@@ -1,5 +1,6 @@
 package net.minecraft.client.network.play;
 
+import com.google.common.base.Charsets;
 import com.google.common.collect.Lists;
 import com.google.common.collect.Maps;
 import com.google.common.collect.Multimap;
@@ -304,8 +305,15 @@
 import net.minecraft.world.chunk.NibbleArray;
 import net.minecraft.world.lighting.WorldLightManager;
 import net.minecraft.world.storage.MapData;
+import net.optifine.reflect.Reflector;
+
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
+import org.vivecraft.api.NetworkHelper;
+import org.vivecraft.api.ServerVivePlayer;
+import org.vivecraft.api.NetworkHelper.PacketDiscriminators;
+import org.vivecraft.provider.MCOpenVR;
+import org.vivecraft.render.PlayerModelController;
 
 public class ClientPlayNetHandler implements IClientPlayNetHandler
 {
@@ -367,6 +375,10 @@
         this.profile = profileIn;
         this.advancementManager = new ClientAdvancementManager(mcIn);
         this.clientSuggestionProvider = new ClientSuggestionProvider(this, mcIn);
+//VIVECRAFT
+        NetworkHelper.resetServerSettings();
+        NetworkHelper.displayedChatMessage = false;
+        //
     }
 
     public ClientSuggestionProvider getSuggestionProvider()
@@ -428,6 +440,9 @@
 
         this.client.debugRenderer.clear();
         this.client.player.preparePlayerToSpawn();
+        //Forge
+        Reflector.call(Reflector.ClientHooks_firePlayerLogin, this.client.playerController, this.client.player, this.client.getConnection().getNetworkManager());
+        //
         int i = packetIn.getPlayerId();
         this.world.addPlayer(i, this.client.player);
         this.client.player.movementInput = new MovementInputFromOptions(this.client.gameSettings);
@@ -439,9 +454,15 @@
         this.client.player.setShowDeathScreen(packetIn.func_229743_k_());
         this.client.playerController.setGameType(packetIn.getGameType());
         this.client.playerController.func_241675_a_(packetIn.func_241786_f_());
+        //Forge
+        Reflector.call(Reflector.NetworkHooks_sendMCRegistryPackets, netManager, "PLAY_TO_SERVER");
+        //
         this.client.gameSettings.sendSettingsToServer();
         this.netManager.sendPacket(new CCustomPayloadPacket(CCustomPayloadPacket.BRAND, (new PacketBuffer(Unpooled.buffer())).writeString(ClientBrandRetriever.getClientModName())));
         this.client.getMinecraftGame().startGameSession();
+        // VIVE START - ask server if it's running this mod
+        NetworkHelper.sendVersionInfo();
+        // VIVE END - ask server if it's running this mod
     }
 
     /**
@@ -961,7 +982,10 @@
 
             if (tileentity != null)
             {
-                tileentity.read(this.world.getBlockState(blockpos), compoundnbt);
+            	if(Reflector.IForgeTileEntity_handleUpdateTag.exists())
+            		Reflector.call(tileentity, Reflector.IForgeTileEntity_handleUpdateTag,this.world.getBlockState(blockpos), compoundnbt);
+            	else
+	                tileentity.read(this.world.getBlockState(blockpos), compoundnbt);
             }
         }
     }
@@ -1077,8 +1101,15 @@
      */
     public void handleChat(SChatPacket packetIn)
     {
-        PacketThreadUtil.checkThreadAndEnqueue(packetIn, this, this.client);
-        this.client.ingameGUI.func_238450_a_(packetIn.getType(), packetIn.getChatComponent(), packetIn.func_240810_e_());
+    	//Forge
+    	PacketThreadUtil.checkThreadAndEnqueue(packetIn, this, this.client);
+    	net.minecraft.util.text.ITextComponent message = (ITextComponent) Reflector.call(Reflector.ForgeEventFactory_onClientChat, packetIn.getType(), packetIn.getChatComponent(), packetIn.func_240810_e_());
+    	if (message == null)     	
+    		this.client.ingameGUI.func_238450_a_(packetIn.getType(), packetIn.getChatComponent(), packetIn.func_240810_e_());
+    	else
+    		this.client.ingameGUI.func_238450_a_(packetIn.getType(), message, packetIn.func_240810_e_());
+		//
+
     }
 
     /**
@@ -1320,6 +1351,10 @@
             this.world.setScoreboard(scoreboard);
             this.client.loadWorld(this.world);
             this.client.displayGuiScreen(new DownloadTerrainScreen());
+            //VIVECRAFT
+            NetworkHelper.resetServerSettings();
+            NetworkHelper.sendVersionInfo();
+            //
         }
 
         this.world.removeAllEntities();
@@ -1344,6 +1379,9 @@
 
         clientplayerentity1.preparePlayerToSpawn();
         clientplayerentity1.setServerBrand(s);
+        //Forge
+        Reflector.call(Reflector.ClientHooks_firePlayerRespawn, this.client.playerController, clientplayerentity, clientplayerentity1, clientplayerentity1.connection.getNetworkManager());
+        //
         this.world.addPlayer(i, clientplayerentity1);
         clientplayerentity1.rotationYaw = -180.0F;
         clientplayerentity1.movementInput = new MovementInputFromOptions(this.client.gameSettings);
@@ -1357,7 +1395,6 @@
         }
 
         this.client.playerController.setGameType(packetIn.getGameType());
-        this.client.playerController.func_241675_a_(packetIn.func_241788_f_());
     }
 
     /**
@@ -1454,7 +1491,7 @@
     {
         PacketThreadUtil.checkThreadAndEnqueue(packetIn, this, this.client);
         Container container = null;
-        PlayerEntity playerentity = this.client.player;
+        ClientPlayerEntity playerentity = this.client.player;
 
         if (packetIn.getWindowId() == 0)
         {
@@ -1469,6 +1506,7 @@
         {
             this.sendPacket(new CConfirmTransactionPacket(packetIn.getWindowId(), packetIn.getActionNumber(), true));
         }
+        
     }
 
     /**
@@ -1521,7 +1559,15 @@
         if (i == 1 && tileentity instanceof MobSpawnerTileEntity || flag || i == 3 && tileentity instanceof BeaconTileEntity || i == 4 && tileentity instanceof SkullTileEntity || i == 6 && tileentity instanceof BannerTileEntity || i == 7 && tileentity instanceof StructureBlockTileEntity || i == 8 && tileentity instanceof EndGatewayTileEntity || i == 9 && tileentity instanceof SignTileEntity || i == 11 && tileentity instanceof BedTileEntity || i == 5 && tileentity instanceof ConduitTileEntity || i == 12 && tileentity instanceof JigsawTileEntity || i == 13 && tileentity instanceof CampfireTileEntity || i == 14 && tileentity instanceof BeehiveTileEntity)
         {
             tileentity.read(this.client.world.getBlockState(blockpos), packetIn.getNbtCompound());
-        }
+        }    		 else {
+    			//Forge
+    			if(tileentity == null) {
+    				LOGGER.error("Received invalid update packet for null tile entity at {} with data: {}", packetIn.getPos(), packetIn.getNbtCompound());
+    				return;
+    			}
+    			Reflector.call(tileentity, Reflector.IForgeTileEntity_onDataPacket, netManager, packetIn);
+    			//
+            }
 
         if (flag && this.client.currentScreen instanceof CommandBlockScreen)
         {
@@ -1783,6 +1829,9 @@
         clientrecipebook.func_243196_a(this.recipeManager.getRecipes());
         clientrecipebook.getRecipes().forEach(imutablesearchtree::func_217872_a);
         imutablesearchtree.recalculate();
+        //Forge
+        Reflector.call(Reflector.ForgeHooksClient_onRecipesUpdated, this.recipeManager);
+        //
     }
 
     public void handlePlayerLook(SPlayerLookPacket packetIn)
@@ -1886,7 +1935,7 @@
 
         if (entity instanceof LivingEntity)
         {
-            Effect effect = Effect.get(packetIn.getEffectId());
+            Effect effect = Effect.get(packetIn.getEffectId() & 0xFF); //Forge FF
 
             if (effect != null)
             {
@@ -1901,7 +1950,22 @@
     {
         PacketThreadUtil.checkThreadAndEnqueue(packetIn, this, this.client);
         ITagCollectionSupplier itagcollectionsupplier = packetIn.getTags();
-        Multimap<ResourceLocation, ResourceLocation> multimap = TagRegistryManager.validateTags(itagcollectionsupplier);
+
+    	//Forge
+    	Multimap<ResourceLocation, ResourceLocation> multimap;
+    	boolean vanillaConnection = false;
+    	if(Reflector.NetworkHooks_isVanillaConnection.exists()) {
+    		vanillaConnection = Reflector.callBoolean(Reflector.NetworkHooks_isVanillaConnection, netManager);
+    		if(vanillaConnection) {
+    			multimap =  TagRegistryManager.validateTags((ITagCollectionSupplier) Reflector.call(Reflector.ForgeTagHandler_withNoCustom, itagcollectionsupplier));
+    		}
+    		else {
+    			multimap = (Multimap<ResourceLocation, ResourceLocation>) Reflector.call(Reflector.TagRegistryManager_validateVanillaTags, itagcollectionsupplier);
+    		}
+    	} else {
+            multimap = TagRegistryManager.validateTags(itagcollectionsupplier);
+    	}       
+        //
 
         if (!multimap.isEmpty())
         {
@@ -1910,6 +1974,13 @@
         }
         else
         {
+        	//Forge
+        	if(Reflector.ForgeTagHandler_resetCachedTagCollections.exists()) {
+        		Reflector.call(Reflector.ForgeTagHandler_resetCachedTagCollections, true, vanillaConnection);
+        		itagcollectionsupplier = (ITagCollectionSupplier) Reflector.call(Reflector.ITagCollectionSupplier_reinjectOptionalTags, itagcollectionsupplier);
+        	}
+        	//
+
             this.networkTagManager = itagcollectionsupplier;
 
             if (!this.netManager.isLocalChannel())
@@ -1917,7 +1988,7 @@
                 itagcollectionsupplier.updateTags();
             }
 
-            this.client.getSearchTree(SearchTreeManager.TAGS).recalculate();
+            this.client.getSearchTree(SearchTreeManager.TAGS).recalculate();     
         }
     }
 
@@ -2280,7 +2351,7 @@
         PacketThreadUtil.checkThreadAndEnqueue(packetIn, this, this.client);
         ResourceLocation resourcelocation = packetIn.getChannelName();
         PacketBuffer packetbuffer = null;
-
+        boolean release = true;
         try
         {
             packetbuffer = packetIn.getBufferData();
@@ -2553,14 +2624,73 @@
                 int l5 = packetbuffer.readInt();
                 this.client.debugRenderer.field_229018_q_.func_229022_a_(blockpos8, l3, s10, l5);
             }
+    		else if (resourcelocation.getNamespace().equalsIgnoreCase("vivecraft"))
+    		{ 			
+    			// VIVE START - server told us that it has this mod too, allow extended reach, etc.
+    			if (resourcelocation.getPath().equalsIgnoreCase("data"))
+    			{ 
+
+    				byte db = packetbuffer.readByte();
+    				PacketDiscriminators dis = PacketDiscriminators.values()[db];
+
+    				switch (dis){
+    				case VERSION:
+    					String v = packetbuffer.readString(1024);
+    					if (!NetworkHelper.displayedChatMessage) {
+    					    NetworkHelper.displayedChatMessage = true;
+    					    this.client.ingameGUI.getChatGUI().printChatMessage(new TranslationTextComponent("vivecraft.messages.serverplugin", v));
+    					}
+    					break;
+    				case REQUESTDATA:
+    					NetworkHelper.serverWantsData = true;
+    					break;
+    				case CLIMBING:
+    					break;
+    				case TELEPORT:
+    					NetworkHelper.serverSupportsDirectTeleport = true;
+    					break;
+    				case UBERPACKET:
+    					Long hi = packetbuffer.readLong();
+    					Long low = packetbuffer.readLong();
+    					byte[] hmd = new byte[29];
+    					byte[] c0 = new byte[29];
+    					byte[] c1 = new byte[29];
+    					packetbuffer.readBytes(29).getBytes(0, hmd);
+    					packetbuffer.readBytes(29).getBytes(0, c0);
+    					packetbuffer.readBytes(29).getBytes(0, c1);
+    					UUID u = new UUID(hi, low);
+
+    					float worldscale = 1f;
+    					float heightscale = 1f;
+    					if (packetbuffer.isReadable())
+    						worldscale = packetbuffer.readFloat();
+    					if (packetbuffer.isReadable())
+    						heightscale = packetbuffer.readFloat();		
+
+    					PlayerModelController.getInstance().Update(u, hmd, c0, c1, worldscale, heightscale);    
+    						
+    					break;
+    				case SETTING_OVERRIDE:
+                        break;
+                    case CRAWL:
+                        NetworkHelper.serverAllowsCrawling = true;
+                        break;
+    				default:
+    					break;
+    				}
+    			}
+    		}
+    		// VIVE END
             else
             {
-                LOGGER.warn("Unknown custom packed identifier: {}", (Object)resourcelocation);
+            	release = false;
+                if (!Reflector.callBoolean(Reflector.NetworkHooks_onCustomPayload, packetIn, this.netManager))               	
+                	LOGGER.warn("Unknown custom packet identifier: {}", (Object)resourcelocation);
             }
         }
         finally
         {
-            if (packetbuffer != null)
+            if (packetbuffer != null && release)
             {
                 packetbuffer.release();
             }
@@ -2955,4 +3085,14 @@
     {
         return this.field_239163_t_;
     }
+    //VIVECRAFT Additions
+    boolean tryParseInt(String value) {  
+    	try {  
+    		Integer.parseInt(value);  
+    		return true;  
+    	} catch (NumberFormatException e) {  
+    		return false;  
+    	}  
+    }
+    //
 }
