diff -r -U 3 minecraft\net\minecraft\world\storage\loot\LootEntryManager.java minecraft_patched\net\minecraft\world\storage\loot\LootEntryManager.java
--- minecraft\net\minecraft\world\storage\loot\LootEntryManager.java
+++ minecraft_patched\net\minecraft\world\storage\loot\LootEntryManager.java
@@ -65,7 +65,7 @@
          if (serializer == null) {
             throw new JsonParseException("Unknown item type: " + p_216193_0_);
          } else {
-            return serializer;
+            return (LootEntry.Serializer<LootEntry>)serializer;
          }
       }
    }

