--- a/net/optifine/reflect/ReflectorForge.java
+++ b/net/optifine/reflect/ReflectorForge.java
@@ -154,7 +154,7 @@
     {
         if (Reflector.IForgeItemStack_isShield.exists())
         {
-            return Reflector.callBoolean(itemstack, Reflector.IForgeItemStack_isShield);
+            return Reflector.callBoolean(itemstack, Reflector.IForgeItemStack_isShield, entityplayer);
         }
         else
         {
