--- a/net/minecraft/resources/ResourcePackList.java
+++ b/net/minecraft/resources/ResourcePackList.java
@@ -4,7 +4,11 @@
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableMap;
 import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Lists;
 import com.google.common.collect.Maps;
+
+import net.minecraft.client.Minecraft;
+
 import java.util.Collection;
 import java.util.List;
 import java.util.Map;
@@ -24,15 +28,23 @@
     public ResourcePackList(ResourcePackInfo.IFactory<T> p_i1984_1_, IPackFinder... p_i1984_2_)
     {
         this.packInfoFactory = p_i1984_1_;
-        this.packFinders = ImmutableSet.copyOf(p_i1984_2_);
+       // this.packFinders = ImmutableSet.copyOf(p_i1984_2_);
+        //Forge
+        this.packFinders = new java.util.HashSet<>(java.util.Arrays.asList(p_i1984_2_));
+        //
     }
 
     public void reloadPacksFromFinders()
     {
-        List<String> list = this.enabled.stream().map(ResourcePackInfo::getName).collect(ImmutableList.toImmutableList());
-        this.close();
-        this.packNameToInfo = this.func_232624_g_();
-        this.enabled = this.func_232618_b_(list);
+    	// Vivecraft
+    	List<T> oldEnabled = Lists.newLinkedList(this.enabled);
+    	List<String> list = this.enabled.stream().map(ResourcePackInfo::getName).collect(ImmutableList.toImmutableList());
+    	this.close();
+    	this.packNameToInfo = this.func_232624_g_();
+    	this.enabled = this.func_232618_b_(list);
+    	// Vivecraft
+    	if (!this.enabled.equals(oldEnabled))
+    		Minecraft.getInstance().resourcePacksChanged = true;
     }
 
     private Map<String, T> func_232624_g_()
@@ -52,7 +64,13 @@
 
     public void setEnabledPacks(Collection<String> p_198985_1_)
     {
-        this.enabled = this.func_232618_b_(p_198985_1_);
+    	// Vivecraft
+    	List<T> oldEnabled = Lists.newLinkedList(this.enabled);
+
+    	this.enabled = this.func_232618_b_(p_198985_1_);
+    	// Vivecraft
+    	if (!this.enabled.equals(oldEnabled))
+    		Minecraft.getInstance().resourcePacksChanged = true;
     }
 
     private List<T> func_232618_b_(Collection<String> p_232618_1_)
@@ -115,4 +133,11 @@
     {
         return this.enabled.stream().map(ResourcePackInfo::getResourcePack).collect(ImmutableList.toImmutableList());
     }
+    
+    //Forge
+    public void addPackFinder(IPackFinder packFinder) {
+        this.packFinders.add(packFinder);
+    }
+    //
+    
 }
