--- a/net/minecraft/client/renderer/ItemRenderer.java
+++ b/net/minecraft/client/renderer/ItemRenderer.java
@@ -12,6 +12,10 @@
 import java.util.Set;
 import javax.annotation.Nullable;
 import net.minecraft.block.Block;
+import org.lwjgl.opengl.GL14;
+import org.vivecraft.gameplay.trackers.SwingTracker;
+import org.vivecraft.gameplay.trackers.TelescopeTracker;
+import org.vivecraft.render.SpecialItemRenderer;
 import net.minecraft.block.BlockState;
 import net.minecraft.block.BreakableBlock;
 import net.minecraft.block.StainedGlassPaneBlock;
@@ -54,6 +58,7 @@
 import net.optifine.EmissiveTextures;
 import net.optifine.reflect.Reflector;
 import net.optifine.reflect.ReflectorForge;
+import net.optifine.render.GlAlphaState;
 import net.optifine.render.VertexBuilderWrapper;
 import net.optifine.shaders.Shaders;
 
@@ -67,6 +72,13 @@
     private final ItemColors itemColors;
     public ModelManager modelManager = null;
     private static boolean renderItemGui = false;
+    
+    //Vivecraft
+    public static boolean ismainhand = false;
+    public static boolean isfphand = false;
+    float fade = 1, manualFade = 1;
+    private GlAlphaState alphaState = new GlAlphaState();
+    //
 
     public ItemRenderer(TextureManager textureManagerIn, ModelManager modelManagerIn, ItemColors itemColorsIn)
     {
@@ -141,6 +153,14 @@
 
             matrixStackIn.translate(-0.5D, -0.5D, -0.5D);
 
+            //VIVECRAFT support held item transparency overrides.
+            ClientPlayerEntity p = Minecraft.getInstance().player;
+            if(p!=null && isfphand){
+            	fade =	SwingTracker.getItemFade(p, itemStackIn);         
+            } else {
+            	fade = manualFade;
+            }
+            //
             if (!modelIn.isBuiltInRenderer() && (itemStackIn.getItem() != Items.TRIDENT || flag))
             {
                 boolean flag1;
@@ -163,6 +183,10 @@
                 {
                     RenderType rendertype = RenderTypeLookup.func_239219_a_(itemStackIn, flag1);
                     IVertexBuilder ivertexbuilder;
+                    //Vivecraft ?\_(?)_/? pick one? make a better one?
+                    if(isfphand && fade < 1)
+                    	rendertype = Atlases.getTranslucentCullBlockType();
+                    //
 
                     if (itemStackIn.getItem() == Items.COMPASS && itemStackIn.hasEffect())
                     {
@@ -333,7 +357,19 @@
                     k = CustomColors.getColorFromItemStack(itemStackIn, bakedquad.getTintIndex(), k);
                 }
             }
-
+            //Vivecraft color overrides for custom items
+            if (Minecraft.getInstance().jumpTracker.isBoots(itemStackIn))
+            {
+            	k = makeColor(1, 0, 255, 0);
+            } else if (Minecraft.getInstance().climbTracker.isClaws(itemStackIn))
+            {
+            	k = makeColor(1, 130, 0, 75);
+            } else if (TelescopeTracker.isTelescope(itemStackIn))
+            {
+            	k = makeColor(1, 190, 110, 135);
+            }
+            //
+		    	
             float f = (float)(k >> 16 & 255) / 255.0F;
             float f1 = (float)(k >> 8 & 255) / 255.0F;
             float f2 = (float)(k & 255) / 255.0F;
@@ -344,7 +380,7 @@
             }
             else
             {
-                bufferIn.addQuad(matrixstack$entry, bakedquad, f, f1, f2, combinedLightIn, combinedOverlayIn);
+       		     SpecialItemRenderer.addQuad(bufferIn, matrixstack$entry, bakedquad, f, f1, f2, fade, combinedLightIn, combinedOverlayIn);
             }
         }
     }
@@ -401,10 +437,15 @@
         this.textureManager.bindTexture(AtlasTexture.LOCATION_BLOCKS_TEXTURE);
         this.textureManager.getTexture(AtlasTexture.LOCATION_BLOCKS_TEXTURE).setBlurMipmapDirect(false, false);
         RenderSystem.enableRescaleNormal();
+        // VIVECRAFT: Save/restore the alpha properly so things stop breaking
+        GlStateManager.getAlphaState(this.alphaState);
+        GlStateManager.enableAlphaTest();
         RenderSystem.enableAlphaTest();
         RenderSystem.defaultAlphaFunc();
         RenderSystem.enableBlend();
-        RenderSystem.blendFunc(GlStateManager.SourceFactor.SRC_ALPHA, GlStateManager.DestFactor.ONE_MINUS_SRC_ALPHA);
+        //RenderSystem.blendFunc(GlStateManager.SourceFactor.SRC_ALPHA, GlStateManager.DestFactor.ONE_MINUS_SRC_ALPHA);
+        // VIVECRAFT: DestFactor of ONE_MINUS_SRC_ALPHA is really what we want here, because math
+        RenderSystem.blendFuncSeparate(GlStateManager.SourceFactor.SRC_ALPHA, GlStateManager.DestFactor.ONE_MINUS_SRC_ALPHA, GlStateManager.SourceFactor.ONE, GlStateManager.DestFactor.ONE_MINUS_SRC_ALPHA);
         RenderSystem.color4f(1.0F, 1.0F, 1.0F, 1.0F);
         RenderSystem.translatef((float)x, (float)y, 100.0F + this.zLevel);
         RenderSystem.translatef(8.0F, 8.0F, 0.0F);
@@ -418,7 +459,6 @@
         {
             RenderHelper.setupGuiFlatDiffuseLighting();
         }
-
         this.renderItem(stack, ItemCameraTransforms.TransformType.GUI, false, matrixstack, irendertypebuffer$impl, 15728880, OverlayTexture.NO_OVERLAY, bakedmodel);
         irendertypebuffer$impl.finish();
         RenderSystem.enableDepthTest();
@@ -588,4 +628,21 @@
     {
         return VanillaResourceType.MODELS;
     }
+    
+    //Vivecraft Additions
+    private int makeColor(int a, int r, int g, int b) {
+    	return  a << 24 | r << 16 | g << 8 | b;
+    }
+    public float getCurrentFade() {
+    	return fade;
+    }
+
+    public float getFade() {
+    	return manualFade;
+    }
+
+    //used for physical inventory.
+    public void setFade(float fade) {
+    	this.manualFade = fade;
+    }
 }
