--- a/net/minecraft/client/renderer/WorldRenderer.java
+++ b/net/minecraft/client/renderer/WorldRenderer.java
@@ -57,6 +57,7 @@
 import net.minecraft.client.renderer.entity.EntityRendererManager;
 import net.minecraft.client.renderer.entity.ItemFrameRenderer;
 import net.minecraft.client.renderer.model.ModelBakery;
+import net.minecraft.client.renderer.model.ModelResourceLocation;
 import net.minecraft.client.renderer.texture.AtlasTexture;
 import net.minecraft.client.renderer.texture.TextureManager;
 import net.minecraft.client.renderer.tileentity.SignTileEntityRenderer;
@@ -68,6 +69,7 @@
 import net.minecraft.client.settings.GraphicsFanciness;
 import net.minecraft.client.settings.ParticleStatus;
 import net.minecraft.client.shader.Framebuffer;
+import net.minecraft.client.shader.FramebufferConstants;
 import net.minecraft.client.shader.ShaderGroup;
 import net.minecraft.client.world.ClientWorld;
 import net.minecraft.client.world.DimensionRenderInfo;
@@ -151,6 +153,11 @@
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 import org.lwjgl.opengl.GL11;
+import org.vivecraft.gameplay.screenhandlers.KeyboardHandler;
+import org.vivecraft.gameplay.screenhandlers.RadialHandler;
+import org.vivecraft.provider.MCOpenVR;
+import org.vivecraft.render.RenderPass;
+import org.vivecraft.settings.VRSettings;
 
 public class WorldRenderer implements IResourceManagerReloadListener, AutoCloseable
 {
@@ -265,7 +272,16 @@
     private static boolean ambientOcclusion = false;
     private Map<String, List<Entity>> mapEntityLists = new HashMap<>();
     private Map<RenderType, Map> mapRegionLayers = new LinkedHashMap<>();
-
+    //Vivecraft
+    @Nullable
+    public Framebuffer alphaSortVROccludedFramebuffer;
+    @Nullable
+    public Framebuffer alphaSortVRUnoccludedFramebuffer;
+    @Nullable 
+    public Framebuffer alphaSortVRHandsFramebuffer;
+    //
+     
+    
     public WorldRenderer(Minecraft mcIn, RenderTypeBuffers rainTimeBuffersIn)
     {
         this.mc = mcIn;
@@ -315,9 +331,16 @@
 
             lightmapIn.enableLightmap();
             World world = this.mc.world;
-            int i = MathHelper.floor(xIn);
-            int j = MathHelper.floor(yIn);
-            int k = MathHelper.floor(zIn);
+            
+            //Vivecraft - fix which block column doesnt get rain.
+            Vector3d me = mc.vrPlayer.vrdata_world_render.getEye(RenderPass.CENTER).getPosition();
+            if(mc.currentPass == RenderPass.THIRD || mc.currentPass == RenderPass.CAMERA)
+                 me = mc.vrPlayer.vrdata_world_render.getEye(mc.currentPass).getPosition();
+            int i = MathHelper.floor(me.x);
+            int j = MathHelper.floor(me.y);
+            int k = MathHelper.floor(me.z);
+            //
+            
             Tessellator tessellator = Tessellator.getInstance();
             BufferBuilder bufferbuilder = tessellator.getBuffer();
             RenderSystem.enableAlphaTest();
@@ -537,14 +560,20 @@
         RenderSystem.texParameter(3553, 10242, 10497);
         RenderSystem.texParameter(3553, 10243, 10497);
         RenderSystem.bindTexture(0);
-        this.makeEntityOutlineShader();
-
-        if (Minecraft.func_238218_y_())
-        {
-            this.func_239233_v_();
-        }
+        
+        //Vivecraft
+        mc.stereoProvider.reinitFrameBuffers("Resource Reload");
+        
+       // this.makeEntityOutlineShader();
+
+//        if (Minecraft.func_238218_y_())
+//        {
+//            this.func_239233_v_();
+//        }
+        //
     }
 
+    //unused
     public void makeEntityOutlineShader()
     {
         if (this.entityOutlineShader != null)
@@ -555,7 +584,7 @@
         ResourceLocation resourcelocation = new ResourceLocation("shaders/post/entity_outline.json");
 
         try
-        {
+        {     	
             this.entityOutlineShader = new ShaderGroup(this.mc.getTextureManager(), this.mc.getResourceManager(), this.mc.getFramebuffer(), resourcelocation);
             this.entityOutlineShader.createBindFramebuffers(this.mc.getMainWindow().getFramebufferWidth(), this.mc.getMainWindow().getFramebufferHeight());
             this.entityOutlineFramebuffer = this.entityOutlineShader.getFramebufferRaw("final");
@@ -573,11 +602,12 @@
             this.entityOutlineFramebuffer = null;
         }
     }
-
-    private void func_239233_v_()
+   
+    //unused make alpha shader
+    public void func_239233_v_()
     {
         this.func_239234_w_();
-        ResourceLocation resourcelocation = new ResourceLocation("shaders/post/transparency.json");
+        ResourceLocation resourcelocation = new ResourceLocation("shaders/post/vrtransparency.json");
 
         try
         {
@@ -629,7 +659,8 @@
         }
     }
 
-    private void func_239234_w_()
+    //unused
+    public void func_239234_w_()
     {
         if (this.field_239227_K_ != null)
         {
@@ -848,14 +879,15 @@
     {
         if (this.world != null)
         {
-            if (Minecraft.func_238218_y_())
-            {
-                this.func_239233_v_();
-            }
-            else
-            {
-                this.func_239234_w_();
-            }
+        	//no no only we can use that word.
+//            if (Minecraft.func_238218_y_())
+//            {
+//                this.func_239233_v_();
+//            }
+//            else
+//            {
+//                this.func_239234_w_();
+//            }
 
             this.world.clearColorCaches();
 
@@ -991,7 +1023,8 @@
             this.viewFrustum.updateChunkPositions(this.mc.player.getPosX(), this.mc.player.getPosZ());
         }
 
-        if (Config.isDynamicLights())
+        // VIVECRAFT: only update on the first frame, otherwise third person makes it go crazy
+        if (Config.isDynamicLights() && mc.currentPass == RenderPass.LEFT)
         {
             DynamicLights.update(this);
         }
@@ -1005,7 +1038,8 @@
         BlockPos blockpos1 = new BlockPos(MathHelper.floor(vector3d.x / 16.0D) * 16, MathHelper.floor(vector3d.y / 16.0D) * 16, MathHelper.floor(vector3d.z / 16.0D) * 16);
         float f = activeRenderInfoIn.getPitch();
         float f1 = activeRenderInfoIn.getYaw();
-        this.displayListEntitiesDirty = this.displayListEntitiesDirty || !this.chunksToUpdate.isEmpty() || vector3d.x != this.lastViewEntityX || vector3d.y != this.lastViewEntityY || vector3d.z != this.lastViewEntityZ || (double)f != this.lastViewEntityPitch || (double)f1 != this.lastViewEntityYaw;
+        // VIVECRAFT: Always update culling because different camera positions
+        this.displayListEntitiesDirty = true; //this.displayListEntitiesDirty || !this.chunksToUpdate.isEmpty() || vector3d.x != this.lastViewEntityX || vector3d.y != this.lastViewEntityY || vector3d.z != this.lastViewEntityZ || (double)f != this.lastViewEntityPitch || (double)f1 != this.lastViewEntityYaw;
         this.lastViewEntityX = vector3d.x;
         this.lastViewEntityY = vector3d.y;
         this.lastViewEntityZ = vector3d.z;
@@ -1349,8 +1383,14 @@
         TileEntityRendererDispatcher.instance.prepare(this.world, this.mc.getTextureManager(), this.mc.fontRenderer, activeRenderInfoIn, this.mc.objectMouseOver);
         this.renderManager.cacheActiveRenderInfo(this.world, activeRenderInfoIn, this.mc.pointedEntity);
         IProfiler iprofiler = this.world.getProfiler();
-        iprofiler.endStartSection("light_updates");
-        this.mc.world.getChunkProvider().getLightManager().tick(Integer.MAX_VALUE, true, true);
+
+    	//Vivecraft once per frame
+    	if(mc.currentPass == RenderPass.LEFT) {
+	        iprofiler.endStartSection("light_updates");
+	        this.mc.world.getChunkProvider().getLightManager().tick(Integer.MAX_VALUE, true, true);
+    	}
+    	//
+    	setShaderGroup();
         Vector3d vector3d = activeRenderInfoIn.getProjectedView();
         double d0 = vector3d.getX();
         double d1 = vector3d.getY();
@@ -1398,7 +1438,12 @@
         {
             Shaders.clearRenderBuffer();
         }
-
+        
+        //Vivecraft - VR Stencil
+        this.mc.getProfiler().endStartSection("stencil");
+    		gameRendererIn.drawEyeStencil(flag1);
+		//    
+       
         if (flag1)
         {
             Shaders.setCamera(matrixStackIn, activeRenderInfoIn, partialTicks);
@@ -1460,12 +1505,13 @@
         iprofiler.endStartSection("terrain");
         Lagometer.timerTerrain.start();
 
-        if (this.mc.gameSettings.ofSmoothFps)
+        // VIVECRAFT: Nope!
+        /*if (this.mc.gameSettings.ofSmoothFps)
         {
             this.mc.getProfiler().endStartSection("finish");
             GL11.glFinish();
             this.mc.getProfiler().endStartSection("terrain");
-        }
+        }*/
 
         if (Config.isFogOff() && FogRenderer.fogStandard)
         {
@@ -1484,14 +1530,16 @@
         }
 
         Lagometer.timerTerrain.end();
-
+        //TODO: fix lighting?
+        Matrix4f light = new Matrix4f();
+        light.setIdentity();
         if (this.world.func_239132_a_().func_239217_c_())
         {
-            RenderHelper.func_237533_a_(matrixStackIn.getLast().getMatrix());
+            RenderHelper.func_237533_a_(light);
         }
         else
         {
-            RenderHelper.setupLevelDiffuseLighting(matrixStackIn.getLast().getMatrix());
+            RenderHelper.setupLevelDiffuseLighting(light);
         }
 
         if (flag1)
@@ -1532,6 +1580,11 @@
             RenderStateManager.enableCache();
         }
 
+        //Vivecraft
+        boolean selfRender = (mc.currentPass == RenderPass.THIRD && mc.vrSettings.displayMirrorMode == VRSettings.MIRROR_THIRD_PERSON) || mc.currentPass == RenderPass.CAMERA;
+        selfRender |= mc.vrSettings.shouldRenderSelf || mc.vrSettings.tmpRenderSelf;
+        //
+
         for (WorldRenderer.LocalRenderInformationContainer worldrenderer$localrenderinformationcontainer : this.renderInfosEntities)
         {
             ChunkRenderDispatcher.ChunkRender chunkrenderdispatcher$chunkrender = worldrenderer$localrenderinformationcontainer.renderChunk;
@@ -1539,7 +1592,8 @@
 
             for (Entity entity : chunk.getEntityLists()[chunkrenderdispatcher$chunkrender.getPosition().getY() / 16])
             {
-                if ((this.renderManager.shouldRender(entity, clippinghelper, d0, d1, d2) || entity.isRidingOrBeingRiddenBy(this.mc.player)) && (entity != activeRenderInfoIn.getRenderViewEntity() || activeRenderInfoIn.isThirdPerson() || activeRenderInfoIn.getRenderViewEntity() instanceof LivingEntity && ((LivingEntity)activeRenderInfoIn.getRenderViewEntity()).isSleeping()) && (!(entity instanceof ClientPlayerEntity) || activeRenderInfoIn.getRenderViewEntity() == entity))
+            	//Vivecraft insert selfRender
+                if ((this.renderManager.shouldRender(entity, clippinghelper, d0, d1, d2) || entity.isRidingOrBeingRiddenBy(this.mc.player)) && (selfRender || entity != activeRenderInfoIn.getRenderViewEntity() || activeRenderInfoIn.isThirdPerson() || activeRenderInfoIn.getRenderViewEntity() instanceof LivingEntity && ((LivingEntity)activeRenderInfoIn.getRenderViewEntity()).isSleeping()) && (!(entity instanceof ClientPlayerEntity) || activeRenderInfoIn.getRenderViewEntity() == entity))
                 {
                     String s = entity.getClass().getName();
                     List<Entity> list = this.mapEntityLists.get(s);
@@ -1561,6 +1615,12 @@
             {
                 ++this.countEntitiesRendered;
 
+                //Vivecraft put back location.
+                if(entity1 == activeRenderInfoIn.getRenderViewEntity()) { 
+                	gameRendererIn.restoreRVEPos((LivingEntity) entity1);
+                }
+                //
+
                 if (entity1.ticksExisted == 0)
                 {
                     entity1.lastTickPosX = entity1.getPosX();
@@ -1596,6 +1656,14 @@
 
                 this.renderEntity(entity1, d0, d1, d2, partialTicks, matrixStackIn, irendertypebuffer);
                 this.renderedEntity = null;
+
+                //Vivecraft
+                if(entity1 == activeRenderInfoIn.getRenderViewEntity()) {
+                	gameRendererIn.cacheRVEPos((LivingEntity) entity1);
+                	gameRendererIn.setupRVE();
+                }
+                //
+
             }
 
             list1.clear();
@@ -1784,10 +1852,25 @@
         --renderEntitiesCounter;
         this.checkMatrixStack(matrixStackIn);
         RayTraceResult raytraceresult = this.mc.objectMouseOver;
-
+        iprofiler.endStartSection("outline");
+        
+        //Vivecraft interact highlight
+        selR = selG = selB = 1f;
+        for (int c=0;c<2;c++) {
+        	if(mc.interactTracker.isInteractActive(c) && (mc.interactTracker.inBlockHit[c] != null || mc.interactTracker.bukkit[c])) {
+        		BlockPos blockpos = mc.interactTracker.inBlockHit[c] != null ? mc.interactTracker.inBlockHit[c].getPos() : new BlockPos(mc.vrPlayer.vrdata_world_render.getController(c).getPosition());
+        		BlockState blockstate = this.world.getBlockState(blockpos);
+        		this.drawSelectionBox(matrixStackIn, irendertypebuffer$impl.getBuffer(RenderType.getLines()), activeRenderInfoIn.getRenderViewEntity(), d0, d1, d2, blockpos, blockstate);
+        		if (c==0) drawBlockOutline = false; //don't draw both
+        	}
+        }
+        //
+             	  
         if (drawBlockOutline && raytraceresult != null && raytraceresult.getType() == RayTraceResult.Type.BLOCK)
         {
-            iprofiler.endStartSection("outline");
+        	//vivecraft put back black outline
+        	selR = selG = selB = 0f;
+        	//
             BlockPos blockpos = ((BlockRayTraceResult)raytraceresult).getPos();
             BlockState blockstate = this.world.getBlockState(blockpos);
             boolean flag6;
@@ -1812,7 +1895,7 @@
                 irendertypebuffer$impl.finish(RenderType.getLines());
             }
         }
-
+   
         RenderSystem.pushMatrix();
         RenderSystem.multMatrix(matrixStackIn.getLast().getMatrix());
         boolean flag5 = GlStateManager.isFogEnabled();
@@ -1820,6 +1903,7 @@
         this.mc.debugRenderer.render(matrixStackIn, irendertypebuffer$impl, d0, d1, d2);
         GlStateManager.setFogEnabled(flag5);
         RenderSystem.popMatrix();
+        this.mc.getProfiler().endStartSection("render bukkake");
         irendertypebuffer$impl.finish(Atlases.getTranslucentCullBlockType());
         irendertypebuffer$impl.finish(Atlases.getBannerType());
         irendertypebuffer$impl.finish(Atlases.getShieldType());
@@ -1836,13 +1920,16 @@
         if (flag1)
         {
             irendertypebuffer$impl.finish();
-            Shaders.preRenderHand();
-            ShadersRender.renderHand0(gameRendererIn, matrixStackIn, activeRenderInfoIn, partialTicks);
+            //Shaders.preRenderHand();
+            //ShadersRender.renderHand0(gameRendererIn, matrixStackIn, activeRenderInfoIn, partialTicks);
             Shaders.preWater();
         }
+        
+        boolean menuHandleft = gameRendererIn.isInMenuRoom() || mc.currentScreen != null || KeyboardHandler.Showing;
+        boolean menuhandright = menuHandleft || (mc.interactTracker.hotbar >= 0 && mc.vrSettings.vrTouchHotbar);
 
         if (this.field_239227_K_ != null)
-        {
+        { //fabulous!
             irendertypebuffer$impl.finish(RenderType.getLines());
             irendertypebuffer$impl.finish();
             this.field_239222_F_.framebufferClear(Minecraft.IS_RUNNING_ON_MAC);
@@ -1856,10 +1943,17 @@
             RenderState.field_239237_T_.setupRenderState();
             iprofiler.endStartSection("particles");
             this.mc.particles.renderParticles(matrixStackIn, irendertypebuffer$impl, lightmapIn, activeRenderInfoIn, partialTicks);
-            RenderState.field_239237_T_.clearRenderState();
+            RenderState.field_239237_T_.clearRenderState();       
+            
+            gameRendererIn.renderVRFabulous(partialTicks, this, menuhandright, menuHandleft);
         }
         else
-        {
+        { //Fast, Fancy
+        	
+        	//Vivecraft do it once.
+            gameRendererIn.renderVrFast(partialTicks, false, menuhandright, menuHandleft);
+            //
+            
             iprofiler.endStartSection("translucent");
 
             if (flag1)
@@ -1886,7 +1980,11 @@
             }
 
             this.mc.particles.renderParticles(matrixStackIn, irendertypebuffer$impl, lightmapIn, activeRenderInfoIn, partialTicks);
-
+        
+        	//Vivecraft do it again. Come at be bro.
+            gameRendererIn.renderVrFast(partialTicks, true, menuhandright, menuHandleft);
+            //
+            
             if (flag1)
             {
                 Shaders.endParticles();
@@ -3192,10 +3290,11 @@
     {
         bufferIn.pos(xIn - camX, (double)yIn - camY, zIn - camZ).tex(texU, texV).endVertex();
     }
-
+    //Vivecraft support colored blockoutline
+    public float selR, selG, selB;
     private void drawSelectionBox(MatrixStack matrixStackIn, IVertexBuilder bufferIn, Entity entityIn, double xIn, double yIn, double zIn, BlockPos blockPosIn, BlockState blockStateIn)
     {
-        drawShape(matrixStackIn, bufferIn, blockStateIn.getShape(this.world, blockPosIn, ISelectionContext.forEntity(entityIn)), (double)blockPosIn.getX() - xIn, (double)blockPosIn.getY() - yIn, (double)blockPosIn.getZ() - zIn, 0.0F, 0.0F, 0.0F, 0.4F);
+        drawShape(matrixStackIn, bufferIn, blockStateIn.getShape(this.world, blockPosIn, ISelectionContext.forEntity(entityIn)), (double)blockPosIn.getX() - xIn, (double)blockPosIn.getY() - yIn, (double)blockPosIn.getZ() - zIn, selR, selG, selB, 0.4F);
     }
 
     public static void drawVoxelShapeParts(MatrixStack matrixStackIn, IVertexBuilder bufferIn, VoxelShape shapeIn, double xIn, double yIn, double zIn, float red, float green, float blue, float alpha)
@@ -3663,6 +3762,8 @@
     {
         Random random = this.world.rand;
 
+        boolean playernear =  mc.player !=null && mc.player.isAlive() && (mc.player.getPosition().distanceSq(blockPosIn) < 25); ///hmm sure why not.
+        
         switch (type)
         {
             case 1000:
@@ -3725,18 +3826,26 @@
                 break;
 
             case 1011:
+                if(playernear)
+                	MCOpenVR.triggerHapticPulse(0,250); //VIVECRAFT go thump.
                 this.world.playSound(blockPosIn, SoundEvents.BLOCK_IRON_DOOR_CLOSE, SoundCategory.BLOCKS, 1.0F, random.nextFloat() * 0.1F + 0.9F, false);
                 break;
 
             case 1012:
+                if(playernear)
+                	MCOpenVR.triggerHapticPulse(0,250); //VIVECRAFT go thump.
                 this.world.playSound(blockPosIn, SoundEvents.BLOCK_WOODEN_DOOR_CLOSE, SoundCategory.BLOCKS, 1.0F, random.nextFloat() * 0.1F + 0.9F, false);
                 break;
 
             case 1013:
+                if(playernear)
+                	MCOpenVR.triggerHapticPulse(0,250); //VIVECRAFT go thump.
                 this.world.playSound(blockPosIn, SoundEvents.BLOCK_WOODEN_TRAPDOOR_CLOSE, SoundCategory.BLOCKS, 1.0F, random.nextFloat() * 0.1F + 0.9F, false);
                 break;
 
             case 1014:
+                if(playernear)
+                	MCOpenVR.triggerHapticPulse(0,250); //VIVECRAFT go thump.
                 this.world.playSound(blockPosIn, SoundEvents.BLOCK_FENCE_GATE_CLOSE, SoundCategory.BLOCKS, 1.0F, random.nextFloat() * 0.1F + 0.9F, false);
                 break;
 
@@ -3757,14 +3866,26 @@
                 break;
 
             case 1019:
+                if(playernear){ //this is just mean.
+                	MCOpenVR.triggerHapticPulse(0,750); 
+                	MCOpenVR.triggerHapticPulse(1,750); 
+                }
                 this.world.playSound(blockPosIn, SoundEvents.ENTITY_ZOMBIE_ATTACK_WOODEN_DOOR, SoundCategory.HOSTILE, 2.0F, (random.nextFloat() - random.nextFloat()) * 0.2F + 1.0F, false);
                 break;
 
             case 1020:
+                if(playernear){ //this is just mean.
+                	MCOpenVR.triggerHapticPulse(0,750); 
+                	MCOpenVR.triggerHapticPulse(1,750); 
+                }
                 this.world.playSound(blockPosIn, SoundEvents.ENTITY_ZOMBIE_ATTACK_IRON_DOOR, SoundCategory.HOSTILE, 2.0F, (random.nextFloat() - random.nextFloat()) * 0.2F + 1.0F, false);
                 break;
 
             case 1021:
+                if(playernear){ //this is just mean.
+                	MCOpenVR.triggerHapticPulse(0,750); 
+                	MCOpenVR.triggerHapticPulse(1,750); 
+                }
                 this.world.playSound(blockPosIn, SoundEvents.ENTITY_ZOMBIE_BREAK_WOODEN_DOOR, SoundCategory.HOSTILE, 2.0F, (random.nextFloat() - random.nextFloat()) * 0.2F + 1.0F, false);
                 break;
 
@@ -3793,10 +3914,17 @@
                 break;
 
             case 1030:
+                if(playernear){ 
+                	MCOpenVR.triggerHapticPulse(0,500); 
+                }
                 this.world.playSound(blockPosIn, SoundEvents.BLOCK_ANVIL_USE, SoundCategory.BLOCKS, 1.0F, random.nextFloat() * 0.1F + 0.9F, false);
                 break;
 
             case 1031:
+                if(playernear){ 
+                	MCOpenVR.triggerHapticPulse(0,1250); 
+                	MCOpenVR.triggerHapticPulse(1,1250); 
+                }
                 this.world.playSound(blockPosIn, SoundEvents.BLOCK_ANVIL_LAND, SoundCategory.BLOCKS, 0.3F, this.world.rand.nextFloat() * 0.1F + 0.9F, false);
                 break;
 
@@ -3817,6 +3945,9 @@
                 break;
 
             case 1036:
+
+                if(playernear)
+                	MCOpenVR.triggerHapticPulse(0,250); //VIVECRAFT go thump.
                 this.world.playSound(blockPosIn, SoundEvents.BLOCK_IRON_TRAPDOOR_CLOSE, SoundCategory.BLOCKS, 1.0F, random.nextFloat() * 0.1F + 0.9F, false);
                 break;
 
@@ -4409,4 +4540,40 @@
             super(p_i232463_1_, p_i232463_2_);
         }
     }
+   
+    
+    //Vivecraft Additions
+    public void setShaderGroup() {
+        this.field_239227_K_ = null;
+        this.field_239222_F_ = null;
+        this.field_239223_G_ = null;
+        this.field_239224_H_ = null;
+        this.field_239225_I_ = null;
+        this.field_239226_J_ = null;
+        this.alphaSortVRHandsFramebuffer = null;
+        this.alphaSortVROccludedFramebuffer = null;
+        this.alphaSortVRUnoccludedFramebuffer = null;
+
+        ShaderGroup shadergroup = this.mc.stereoProvider.alphaShaders.get(mc.getFramebuffer().name);
+        if (shadergroup != null) { 
+	        this.field_239227_K_ = shadergroup;
+	        this.field_239222_F_ = shadergroup.getFramebufferRaw("translucent");
+	        this.field_239223_G_ = shadergroup.getFramebufferRaw("itemEntity");
+	        this.field_239224_H_ = shadergroup.getFramebufferRaw("particles");
+	        this.field_239225_I_ = shadergroup.getFramebufferRaw("weather");
+	        this.field_239226_J_ = shadergroup.getFramebufferRaw("clouds");
+	        this.alphaSortVRHandsFramebuffer = shadergroup.getFramebufferRaw("vrhands");
+	        this.alphaSortVROccludedFramebuffer = shadergroup.getFramebufferRaw("vroccluded");
+	        this.alphaSortVRUnoccludedFramebuffer = shadergroup.getFramebufferRaw("vrunoccluded");
+        } 
+        
+        this.entityOutlineShader = null;
+        this.entityOutlineFramebuffer = null;
+        shadergroup =  this.mc.stereoProvider.entityShaders.get(mc.getFramebuffer().name);
+        if (shadergroup != null) {
+            this.entityOutlineShader = shadergroup;
+            this.entityOutlineFramebuffer = this.entityOutlineShader.getFramebufferRaw("final");
+        }  
+
+    }
 }
