--- a/net/minecraft/fluid/FluidState.java
+++ b/net/minecraft/fluid/FluidState.java
@@ -18,7 +18,8 @@
 import net.minecraft.world.IBlockReader;
 import net.minecraft.world.World;
 
-public final class FluidState extends StateHolder<Fluid, FluidState>
+// VIVECRAFT: Why the bull fucking hell is this final? I'm going to shit a hotel.
+public class FluidState extends StateHolder<Fluid, FluidState>
 {
     public static final Codec<FluidState> field_237213_a_ = func_235897_a_(Registry.FLUID, Fluid::getDefaultState).stable();
 
