--- a/net/minecraft/client/gui/screen/ChatScreen.java
+++ b/net/minecraft/client/gui/screen/ChatScreen.java
@@ -134,7 +134,14 @@
                 this.sendMessage(s);
             }
 
-            this.minecraft.displayGuiScreen((Screen)null);
+            // VIVECRAFT: don't close the chat in standing mode
+            if (this.minecraft.vrSettings.seated) {
+                this.minecraft.displayGuiScreen((Screen)null);
+            } else {
+                this.inputField.setText(this.defaultInputFieldText);
+                this.sentHistoryCursor = this.minecraft.ingameGUI.getChatGUI().getSentMessages().size();
+            }
+            //
             return true;
         }
     }
