--- a/net/minecraft/client/gui/screen/EditSignScreen.java
+++ b/net/minecraft/client/gui/screen/EditSignScreen.java
@@ -5,6 +5,9 @@
 import com.mojang.blaze3d.systems.RenderSystem;
 import com.mojang.blaze3d.vertex.IVertexBuilder;
 import java.util.List;
+
+import org.vivecraft.gameplay.screenhandlers.KeyboardHandler;
+
 import net.minecraft.block.BlockState;
 import net.minecraft.block.StandingSignBlock;
 import net.minecraft.client.gui.RenderComponentsUtil;
@@ -50,6 +53,7 @@
             this.close();
         }));
         this.tileSign.setEditable(false);
+        KeyboardHandler.setOverlayShowing(true);
         this.textInputUtil = new TextInputUtil(this.minecraft, () ->
         {
             return this.tileSign.getText(this.editLine).getString();
@@ -70,6 +74,7 @@
         }
 
         this.tileSign.setEditable(true);
+        KeyboardHandler.setOverlayShowing(false);
     }
 
     public void tick()
