--- a/net/minecraft/client/gui/screen/EditBookScreen.java
+++ b/net/minecraft/client/gui/screen/EditBookScreen.java
@@ -41,6 +41,7 @@
 import org.apache.commons.lang3.StringUtils;
 import org.apache.commons.lang3.mutable.MutableBoolean;
 import org.apache.commons.lang3.mutable.MutableInt;
+import org.vivecraft.gameplay.screenhandlers.KeyboardHandler;
 
 public class EditBookScreen extends Screen
 {
@@ -176,6 +177,9 @@
             this.previousPage();
         }, true));
         this.updateButtons();
+        //Vivecraft
+        KeyboardHandler.setOverlayShowing(true);
+        //
     }
 
     private void previousPage()
