--- a/net/minecraft/client/gui/screen/OptionsScreen.java
+++ b/net/minecraft/client/gui/screen/OptionsScreen.java
@@ -3,6 +3,9 @@
 import com.google.common.collect.ImmutableList;
 import com.mojang.blaze3d.matrix.MatrixStack;
 import java.util.List;
+
+import org.vivecraft.gui.settings.GuiMainVRSettings;
+
 import net.minecraft.client.AbstractOption;
 import net.minecraft.client.GameSettings;
 import net.minecraft.client.gui.AccessibilityScreen;
@@ -15,6 +18,7 @@
 import net.minecraft.resources.ResourcePackInfo;
 import net.minecraft.resources.ResourcePackList;
 import net.minecraft.util.text.ITextComponent;
+import net.minecraft.util.text.StringTextComponent;
 import net.minecraft.util.text.TranslationTextComponent;
 import net.minecraft.world.Difficulty;
 
@@ -42,7 +46,17 @@
         {
             int j = this.width / 2 - 155 + i % 2 * 160;
             int k = this.height / 6 - 12 + 24 * (i >> 1);
-            this.addButton(abstractoption.createWidget(this.minecraft.gameSettings, j, k, 150));
+            //Vivecraft
+            if (abstractoption == abstractoption.FOV) {
+            	this.addButton(new Button(j, k, 150, 20, new TranslationTextComponent("vivecraft.options.screen.main.button"), (p) ->
+            	{
+            		minecraft.gameSettings.saveOptions();
+            		minecraft.displayGuiScreen(new GuiMainVRSettings(OptionsScreen.this));
+            	}));
+            } else {
+                this.addButton(abstractoption.createWidget(this.minecraft.gameSettings, j, k, 150));
+            }
+            
             ++i;
         }
 
