--- a/net/minecraft/client/gui/screen/OptionsSoundsScreen.java
+++ b/net/minecraft/client/gui/screen/OptionsSoundsScreen.java
@@ -8,7 +8,9 @@
 import net.minecraft.client.gui.widget.button.Button;
 import net.minecraft.client.gui.widget.button.OptionButton;
 import net.minecraft.util.SoundCategory;
+import net.minecraft.util.text.StringTextComponent;
 import net.minecraft.util.text.TranslationTextComponent;
+import org.vivecraft.settings.VRSettings;
 
 public class OptionsSoundsScreen extends SettingsScreen
 {
@@ -31,6 +33,17 @@
                 ++i;
             }
         }
+        
+        //Vivecraft
+        this.addButton(new Button(this.width / 2 - 155 + i % 2 * 160, this.height / 6 -12 + 24 * (i >>1), 150, 20, new StringTextComponent(minecraft.vrSettings.getButtonDisplayString(VRSettings.VrOptions.HRTF_SELECTION)), (p_213104_1_) ->
+        {
+        	this.buttons.clear();
+        	this.children.clear();
+        	minecraft.vrSettings.setOptionValue(VRSettings.VrOptions.HRTF_SELECTION);
+        	minecraft.vrSettings.saveOptions();
+        	this.init();
+        }));
+        ///
 
         int j = this.width / 2 - 75;
         int k = this.height / 6 - 12;
