--- a/net/minecraft/client/gui/screen/OptionsScreen.java
+++ b/net/minecraft/client/gui/screen/OptionsScreen.java
@@ -1,5 +1,7 @@
 package net.minecraft.client.gui.screen;
 
+import org.vivecraft.gui.settings.GuiMainVRSettings;
+
 import net.minecraft.client.GameSettings;
 import net.minecraft.client.gui.AccessibilityScreen;
 import net.minecraft.client.gui.widget.button.Button;
@@ -36,7 +38,17 @@
         {
             int j = this.width / 2 - 155 + i % 2 * 160;
             int k = this.height / 6 - 12 + 24 * (i >> 1);
-            this.addButton(abstractoption.createWidget(this.minecraft.gameSettings, j, k, 150));
+            //Vivecraft
+            if (abstractoption == abstractoption.FOV) {
+            	this.addButton(new Button(j, k, 150, 20, "VR Settings...", (p) ->   
+            	{
+            		minecraft.gameSettings.saveOptions();
+            		minecraft.displayGuiScreen(new GuiMainVRSettings(OptionsScreen.this));
+            	}));
+            } else {
+                this.addButton(abstractoption.createWidget(this.minecraft.gameSettings, j, k, 150));
+            }
+            
             ++i;
         }
 
