--- a/net/minecraft/client/gui/screen/Screen.java
+++ b/net/minecraft/client/gui/screen/Screen.java
@@ -38,6 +38,8 @@
 import net.minecraft.util.text.Style;
 import net.minecraft.util.text.event.ClickEvent;
 import net.minecraft.util.text.event.HoverEvent;
+import net.optifine.reflect.Reflector;
+
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
@@ -128,6 +130,12 @@
 
     protected void renderTooltip(MatrixStack matrixStack, ItemStack itemStack, int mouseX, int mouseY)
     {
+    	//Forge TODO fix
+    	Reflector.call(Reflector.GuiUtils_preItemToolTip, itemStack);        
+    	FontRenderer fon = (FontRenderer)Reflector.call(itemStack.getItem(), Reflector.IForgeItem_getFontRenderer, itemStack);
+        this.renderToolTip(matrixStack, Lists.transform(this.getTooltipFromItem(itemStack), ITextComponent::func_241878_f), mouseX, mouseY, (font == null ? this.font : font));
+    	Reflector.call(Reflector.GuiUtils_postItemToolTip);
+    	//
         this.func_243308_b(matrixStack, this.getTooltipFromItem(itemStack), mouseX, mouseY);
     }
 
@@ -146,7 +154,15 @@
         this.renderTooltip(p_243308_1_, Lists.transform(p_243308_2_, ITextComponent::func_241878_f), p_243308_3_, p_243308_4_);
     }
 
-    public void renderTooltip(MatrixStack matrixStack, List <? extends IReorderingProcessor > tooltips, int mouseX, int mouseY)
+    //Forge vanilla method
+    public void renderTooltip(MatrixStack p_238654_1_, List<? extends IReorderingProcessor> tooltips, int mouseX, int mouseY) {
+        this.renderToolTip(p_238654_1_, tooltips, mouseX, mouseY, font);
+     }
+    //
+    
+
+    //Captial T -.-
+    public void renderToolTip(MatrixStack matrixStack, List <? extends IReorderingProcessor > tooltips, int mouseX, int mouseY, FontRenderer font)
     {
         if (!tooltips.isEmpty())
         {
@@ -365,6 +381,11 @@
 
     public void sendMessage(String text, boolean addToChat)
     {
+    	//Forge
+    	if(Reflector.ForgeEventFactory_onClientSendMessage.exists())
+    		text = Reflector.callString(Reflector.ForgeEventFactory_onClientSendMessage, text);
+        if (text.isEmpty()) return;
+        //
         if (addToChat)
         {
             this.minecraft.ingameGUI.getChatGUI().addToSentMessages(text);
@@ -380,10 +401,15 @@
         this.font = minecraft.fontRenderer;
         this.width = width;
         this.height = height;
-        this.buttons.clear();
-        this.children.clear();
-        this.setListener((IGuiEventListener)null);
-        this.init();
+        java.util.function.Consumer<Widget> remove = (b) -> { buttons.remove(b); children.remove(b); };
+        java.util.function.Consumer<Widget> add = (b) -> { this.addButton(b); };
+        if (!Reflector.postForgeBusEvent(Reflector.InitGuiEvent_Pre_Constructor, this, this.buttons, add, remove)) {
+	        this.buttons.clear();
+	        this.children.clear();
+	        this.setListener((IGuiEventListener)null);
+	        this.init();
+        }
+        Reflector.postForgeBusEvent(Reflector.InitGuiEvent_Post_Constructor, this, this.buttons, add, remove);
     }
 
     public List <? extends IGuiEventListener > getEventListeners()
@@ -412,7 +438,17 @@
     {
         if (this.minecraft.world != null)
         {
-            this.fillGradient(matrixStack, 0, 0, this.width, this.height, -1072689136, -804253680);
+        	//Vivecraft
+        	if (minecraft.vrSettings != null && minecraft.vrSettings.menuBackground == false)
+        	{
+        		this.fillGradient(matrixStack, 0, 0, this.width, this.height, 0, 0);
+        	}
+        	else
+        	{
+        		this.fillGradient(matrixStack, 0, 0, this.width, this.height, -1072689136, -804253680);
+        	}
+        	//     	
+            Reflector.postForgeBusEvent(Reflector.BackgroundDrawnEvent_Constructor, this, matrixStack);
         }
         else
         {
@@ -433,6 +469,7 @@
         bufferbuilder.pos((double)this.width, 0.0D, 0.0D).tex((float)this.width / 32.0F, (float)vOffset).color(64, 64, 64, 255).endVertex();
         bufferbuilder.pos(0.0D, 0.0D, 0.0D).tex(0.0F, (float)vOffset).color(64, 64, 64, 255).endVertex();
         tessellator.draw();
+        Reflector.postForgeBusEvent(Reflector.BackgroundDrawnEvent_Constructor, this, new MatrixStack());
     }
 
     public boolean isPauseScreen()
@@ -548,4 +585,10 @@
     public void addPacks(List<Path> packs)
     {
     }
+    
+    //Forge
+    public Minecraft getMinecraft() {
+    	return this.minecraft;
+    }
+    //
 }
