--- a/net/minecraft/client/gui/screen/Screen.java
+++ b/net/minecraft/client/gui/screen/Screen.java
@@ -38,6 +38,8 @@
 import net.minecraft.util.text.Style;
 import net.minecraft.util.text.event.ClickEvent;
 import net.minecraft.util.text.event.HoverEvent;
+import net.optifine.reflect.Reflector;
+
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
@@ -360,6 +362,11 @@
 
     public void sendMessage(String p_231159_1_, boolean p_231159_2_)
     {
+    	//Forge
+    	if(Reflector.ForgeEventFactory_onClientSendMessage.exists())
+    		p_231159_1_ = Reflector.callString(Reflector.ForgeEventFactory_onClientSendMessage, p_231159_1_);
+        if (p_231159_1_.isEmpty()) return;
+        //
         if (p_231159_2_)
         {
             this.minecraft.ingameGUI.getChatGUI().addToSentMessages(p_231159_1_);
@@ -428,6 +435,7 @@
         bufferbuilder.pos((double)this.width, 0.0D, 0.0D).tex((float)this.width / 32.0F, (float)p_231165_1_).color(64, 64, 64, 255).endVertex();
         bufferbuilder.pos(0.0D, 0.0D, 0.0D).tex(0.0F, (float)p_231165_1_).color(64, 64, 64, 255).endVertex();
         tessellator.draw();
+        Reflector.postForgeBusEvent(Reflector.BackgroundDrawnEvent_Constructor, this);
     }
 
     public boolean isPauseScreen()
@@ -543,4 +551,9 @@
     public void func_230476_a_(List<Path> p_230476_1_)
     {
     }
+    
+    public Minecraft getMinecraft() {
+        return this.minecraft;
+     }
+    
 }
