--- a/net/minecraft/client/renderer/entity/EntityRenderer.java
+++ b/net/minecraft/client/renderer/entity/EntityRenderer.java
@@ -135,7 +135,9 @@
             int i = "deadmau5".equals(displayNameIn.getString()) ? -10 : 0;
             matrixStackIn.push();
             matrixStackIn.translate(0.0D, (double)f, 0.0D);
-            matrixStackIn.rotate(this.renderManager.getCameraOrientation());
+            //Vivecraft get correct normal for nameplate
+            matrixStackIn.rotate(this.renderManager.getCameraOrientationOffset(0.5f));
+            //
             matrixStackIn.scale(-0.025F, -0.025F, 0.025F);
             Matrix4f matrix4f = matrixStackIn.getLast().getMatrix();
             float f1 = Minecraft.getInstance().gameSettings.getTextBackgroundOpacity(0.25F);
