--- a/net/minecraft/client/renderer/entity/EntityRendererManager.java
+++ b/net/minecraft/client/renderer/entity/EntityRendererManager.java
@@ -7,6 +7,10 @@
 import java.util.Collections;
 import java.util.Map;
 import javax.annotation.Nullable;
+
+import org.vivecraft.render.PlayerModelController;
+import org.vivecraft.render.PlayerModelController.RotInfo;
+
 import net.minecraft.block.BlockRenderType;
 import net.minecraft.block.BlockState;
 import net.minecraft.client.GameSettings;
@@ -58,6 +62,10 @@
     private final Map<EntityType, EntityRenderer> renderers = Maps.newHashMap();
     private final Map<String, PlayerRenderer> skinMap = Maps.newHashMap();
     private final PlayerRenderer playerRenderer;
+	//VIVECRAFT
+	private final Map<String, VRPlayerRenderer> skinMapVR = Maps.<String, VRPlayerRenderer>newHashMap();
+	private VRPlayerRenderer playerRendererVR;
+	//
     private final FontRenderer textRenderer;
     public final TextureManager textureManager;
     private World world;
@@ -219,9 +227,24 @@
     {
         if (entityIn instanceof AbstractClientPlayerEntity)
         {
-            String s = ((AbstractClientPlayerEntity)entityIn).getSkinType();
-            PlayerRenderer playerrenderer = this.skinMap.get(s);
-            return (EntityRenderer <? super T >)(playerrenderer != null ? playerrenderer : this.playerRenderer);
+        	String s = ((AbstractClientPlayerEntity)entityIn).getSkinType();
+        	PlayerRenderer playerrenderer = this.skinMap.get(s);
+        	//VIVECRAFT
+
+        	RotInfo rotinfo = PlayerModelController.getInstance().getRotationsForPlayer(entityIn.getUniqueID());
+        	if(rotinfo!=null && !rotinfo.seated)
+        	{
+        		if (this.playerRendererVR == null) { // Needs to lazy load cause RenderGlobal
+        			this.playerRendererVR = new VRPlayerRenderer(this);
+        			this.skinMapVR.put("default", this.playerRendererVR);
+        			this.skinMapVR.put("slim", new VRPlayerRenderer(this, true));
+        		}
+        		VRPlayerRenderer renderplayervr = this.skinMapVR.get(s);
+
+        		return (EntityRenderer <? super T >) (renderplayervr != null ? renderplayervr : this.playerRendererVR);            	
+        	}  else 
+        		return (EntityRenderer <? super T >) (playerrenderer != null ? playerrenderer : this.playerRenderer);
+        	//
         }
         else
         {
