--- a/net/minecraft/client/renderer/entity/FishRenderer.java
+++ b/net/minecraft/client/renderer/entity/FishRenderer.java
@@ -1,5 +1,7 @@
 package net.minecraft.client.renderer.entity;
 
+import org.vivecraft.provider.openvr_jna.MCOpenVR;
+
 import com.mojang.blaze3d.matrix.MatrixStack;
 import com.mojang.blaze3d.vertex.IVertexBuilder;
 import net.minecraft.client.Minecraft;
@@ -51,7 +53,7 @@
             int i = playerentity.getPrimaryHand() == HandSide.RIGHT ? 1 : -1;
             ItemStack itemstack = playerentity.getHeldItemMainhand();
 
-            if (itemstack.getItem() != Items.FISHING_ROD)
+            if (!(itemstack.getItem() instanceof net.minecraft.item.FishingRodItem))
             {
                 i = -i;
             }
@@ -81,6 +83,18 @@
                 d5 = MathHelper.lerp((double)partialTicks, playerentity.prevPosY, playerentity.getPosY()) + vector3d.y;
                 d6 = MathHelper.lerp((double)partialTicks, playerentity.prevPosZ, playerentity.getPosZ()) + vector3d.z;
                 f3 = playerentity.getEyeHeight();
+                
+                // VIVE START - use controller for fishing rod start point
+            	int c = 1;
+            	if (playerentity.getHeldItemMainhand().getItem() instanceof net.minecraft.item.FishingRodItem) c = 0;
+            	Vector3d aimSource = Minecraft.getInstance().gameRenderer.getControllerRenderPos(c); 
+            	Vector3d aimRotation = Minecraft.getInstance().vrPlayer.vrdata_world_render.getHand(c).getDirection();
+            	d4= aimSource.x + aimRotation.x * 0.47f;
+            	d5= aimSource.y + aimRotation.y * 0.47f;
+            	d6 = aimSource.z + aimRotation.z * 0.47f;
+            	f3=0;
+            	// VIVE END
+            	
             }
             else
             {
