--- a/net/minecraft/client/renderer/entity/GuardianRenderer.java
+++ b/net/minecraft/client/renderer/entity/GuardianRenderer.java
@@ -1,7 +1,11 @@
 package net.minecraft.client.renderer.entity;
 
+import org.vivecraft.render.RenderPass;
+
 import com.mojang.blaze3d.matrix.MatrixStack;
 import com.mojang.blaze3d.vertex.IVertexBuilder;
+
+import net.minecraft.client.Minecraft;
 import net.minecraft.client.renderer.IRenderTypeBuffer;
 import net.minecraft.client.renderer.RenderType;
 import net.minecraft.client.renderer.culling.ClippingHelper;
@@ -78,7 +82,13 @@
             float f3 = entityIn.getEyeHeight();
             matrixStackIn.push();
             matrixStackIn.translate(0.0D, (double)f3, 0.0D);
+           
+            //Vivecraft
             Vector3d vector3d = this.getPosition(livingentity, (double)livingentity.getHeight() * 0.5D, partialTicks);
+            if(livingentity == Minecraft.getInstance().getRenderViewEntity())
+                vector3d = Minecraft.getInstance().vrPlayer.getVRDataWorld().getEye(RenderPass.CENTER).getPosition().subtract(0,0.3 * Minecraft.getInstance().vrPlayer.worldScale,0);
+            //         
+
             Vector3d vector3d1 = this.getPosition(entityIn, (double)f3, partialTicks);
             Vector3d vector3d2 = vector3d.subtract(vector3d1);
             float f4 = (float)(vector3d2.length() + 1.0D);
