--- a/net/minecraft/client/renderer/entity/MobRenderer.java
+++ b/net/minecraft/client/renderer/entity/MobRenderer.java
@@ -2,6 +2,8 @@
 
 import com.mojang.blaze3d.matrix.MatrixStack;
 import com.mojang.blaze3d.vertex.IVertexBuilder;
+
+import net.minecraft.client.Minecraft;
 import net.minecraft.client.renderer.IRenderTypeBuffer;
 import net.minecraft.client.renderer.LightTexture;
 import net.minecraft.client.renderer.RenderType;
@@ -77,6 +79,15 @@
             double d7 = MathHelper.lerp((double)partialTicks, leashHolder.prevPosY + (double)leashHolder.getEyeHeight() * 0.7D, leashHolder.getPosY() + (double)leashHolder.getEyeHeight() * 0.7D) - d4 * 0.5D - 0.25D;
             double d8 = MathHelper.lerp((double)partialTicks, leashHolder.prevPosZ, leashHolder.getPosZ()) - d3 * 0.7D + d2 * 0.5D * d5;
             double d9 = (double)(MathHelper.lerp(partialTicks, entityLivingIn.renderYawOffset, entityLivingIn.prevRenderYawOffset) * ((float)Math.PI / 180F)) + (Math.PI / 2D);
+            // VIVE START 
+            if (leashHolder == Minecraft.getInstance().player)
+            {
+            	Vector3d aimSource = Minecraft.getInstance().vrPlayer.vrdata_world_render.getController(0).getPosition();
+            	d6 = aimSource.x;
+            	d7 = aimSource.y; 
+            	d8 = aimSource.z;             		         
+            }
+            // VIVE END 
             Vector3d vector3d = entityLivingIn.func_241205_ce_();
             d2 = Math.cos(d9) * vector3d.z + Math.sin(d9) * vector3d.x;
             d3 = Math.sin(d9) * vector3d.z - Math.cos(d9) * vector3d.x;
