--- a/net/minecraft/client/renderer/entity/MobRenderer.java
+++ b/net/minecraft/client/renderer/entity/MobRenderer.java
@@ -2,6 +2,8 @@
 
 import com.mojang.blaze3d.matrix.MatrixStack;
 import com.mojang.blaze3d.vertex.IVertexBuilder;
+
+import net.minecraft.client.Minecraft;
 import net.minecraft.client.renderer.IRenderTypeBuffer;
 import net.minecraft.client.renderer.LightTexture;
 import net.minecraft.client.renderer.RenderType;
