--- a/net/minecraft/client/gui/widget/TextFieldWidget.java
+++ b/net/minecraft/client/gui/widget/TextFieldWidget.java
@@ -8,12 +8,16 @@
 import java.util.function.Consumer;
 import java.util.function.Predicate;
 import javax.annotation.Nullable;
+
+import org.vivecraft.gameplay.screenhandlers.KeyboardHandler;
+
 import net.minecraft.client.Minecraft;
 import net.minecraft.client.gui.AbstractGui;
 import net.minecraft.client.gui.FontRenderer;
 import net.minecraft.client.gui.IGuiEventListener;
 import net.minecraft.client.gui.IRenderable;
 import net.minecraft.client.gui.screen.Screen;
+import net.minecraft.client.gui.screen.SleepInMultiplayerScreen;
 import net.minecraft.client.renderer.BufferBuilder;
 import net.minecraft.client.renderer.Tessellator;
 import net.minecraft.client.renderer.vertex.DefaultVertexFormats;
@@ -449,6 +453,10 @@
                 this.setFocused2(flag);
             }
 
+            // VIVE START - show keyboard
+            if(flag) 
+            	KeyboardHandler.setOverlayShowing(true);       
+            // VIVE END - show keyboard
             if (this.isFocused() && flag && p_231044_5_ == 0)
             {
                 int i = MathHelper.floor(p_231044_1_) - this.x;
@@ -504,7 +512,12 @@
             {
                 String s1 = flag ? s.substring(0, j) : s;
                 j1 = this.fontRenderer.func_238405_a_(p_230431_1_, this.textFormatter.apply(s1, this.lineScrollOffset), (float)l, (float)i1, i2);
+                //VIVECRAFT
+            } else {
+            	if(!Minecraft.getInstance().vrSettings.seated && !KeyboardHandler.Showing)
+            		this.fontRenderer.drawStringWithShadow("Click to open keyboard", (float)l, (float)i1, this.disabledColor);
             }
+            //
 
             boolean flag2 = this.cursorPosition < this.text.length() || this.text.length() >= this.getMaxStringLength();
             int k1 = j1;
@@ -647,6 +660,12 @@
         if (p_230995_1_)
         {
             this.cursorCounter = 0;
+            
+        	// VIVE START - show keyboard
+        	if(Minecraft.getInstance().currentScreen instanceof SleepInMultiplayerScreen) return;
+        	if(Minecraft.getInstance().vrSettings.autoOpenKeyboard)
+        		KeyboardHandler.setOverlayShowing(true);
+        	// VIVE END - show keyboard
         }
     }
 
@@ -701,7 +720,8 @@
 
     public void setCanLoseFocus(boolean canLoseFocusIn)
     {
-        this.canLoseFocus = canLoseFocusIn;
+    	//Vivecraft - I don't care what you think.
+    	this.canLoseFocus = false;    
     }
 
     public boolean getVisible()
