--- a/net/minecraft/client/renderer/model/ModelBakery.java
+++ b/net/minecraft/client/renderer/model/ModelBakery.java
@@ -39,8 +39,10 @@
 import net.minecraft.block.BlockRenderType;
 import net.minecraft.block.BlockState;
 import net.minecraft.block.Blocks;
+import net.minecraft.client.Minecraft;
 import net.minecraft.client.renderer.Atlases;
 import net.minecraft.client.renderer.BlockModelShapes;
+import net.minecraft.client.renderer.GameRenderer;
 import net.minecraft.client.renderer.RenderType;
 import net.minecraft.client.renderer.color.BlockColors;
 import net.minecraft.client.renderer.model.multipart.Multipart;
@@ -71,6 +73,7 @@
 import org.apache.commons.lang3.tuple.Triple;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
+import org.vivecraft.gameplay.trackers.TelescopeTracker;
 
 public class ModelBakery
 {
@@ -203,7 +206,13 @@
                 this.loadTopModel(BlockModelShapes.getModelLocation(p_lambda$processLoading$8_1_));
             });
         }
-
+        
+        //Vivecraft custom models
+		this.loadTopModel(TelescopeTracker.scopeModel);
+        this.loadTopModel(GameRenderer.thirdPersonCameraModel);
+        this.loadTopModel(GameRenderer.thirdPersonCameraDisplayModel);
+        //
+		
         p_processLoading_1_.endStartSection("items");
 
         for (ResourceLocation resourcelocation : Registry.ITEM.keySet())
