--- a/net/minecraft/client/renderer/model/ModelBakery.java
+++ b/net/minecraft/client/renderer/model/ModelBakery.java
@@ -39,8 +39,10 @@
 import net.minecraft.block.BlockRenderType;
 import net.minecraft.block.BlockState;
 import net.minecraft.block.Blocks;
+import net.minecraft.client.Minecraft;
 import net.minecraft.client.renderer.Atlases;
 import net.minecraft.client.renderer.BlockModelShapes;
+import net.minecraft.client.renderer.GameRenderer;
 import net.minecraft.client.renderer.RenderType;
 import net.minecraft.client.renderer.color.BlockColors;
 import net.minecraft.client.renderer.model.multipart.Multipart;
@@ -72,6 +74,9 @@
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
+import org.vivecraft.gameplay.trackers.CameraTracker;
+import org.vivecraft.gameplay.trackers.TelescopeTracker;
+
 public class ModelBakery
 {
     public static final RenderMaterial LOCATION_FIRE_0 = new RenderMaterial(AtlasTexture.LOCATION_BLOCKS_TEXTURE, new ResourceLocation("block/fire_0"));
@@ -203,7 +208,15 @@
                 this.loadTopModel(BlockModelShapes.getModelLocation(p_lambda$processLoading$8_1_));
             });
         }
-
+        
+        //Vivecraft custom models
+		this.loadTopModel(TelescopeTracker.scopeModel);
+        this.loadTopModel(GameRenderer.thirdPersonCameraModel);
+        this.loadTopModel(GameRenderer.thirdPersonCameraDisplayModel);
+        this.loadTopModel(CameraTracker.cameraModel);
+        this.loadTopModel(CameraTracker.cameraDisplayModel);
+        //
+		
         p_processLoading_1_.endStartSection("items");
 
         for (ResourceLocation resourcelocation : Registry.ITEM.keySet())
