--- a/net/minecraft/client/renderer/model/ModelBakery.java
+++ b/net/minecraft/client/renderer/model/ModelBakery.java
@@ -71,6 +71,7 @@
 import org.apache.commons.lang3.tuple.Triple;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
+import org.vivecraft.gameplay.trackers.TelescopeTracker;
 
 public class ModelBakery
 {
@@ -202,7 +203,11 @@
                 this.loadTopModel(BlockModelShapes.getModelLocation(p_lambda$processLoading$8_1_));
             });
         }
-
+        
+        //Vivecraft custom models
+		this.loadTopModel(TelescopeTracker.scopeModel);
+        //
+		
         p_processLoading_1_.endStartSection("items");
 
         for (ResourceLocation resourcelocation : Registry.ITEM.keySet())
