--- a/net/minecraft/item/CrossbowItem.java
+++ b/net/minecraft/item/CrossbowItem.java
@@ -5,6 +5,11 @@
 import java.util.Random;
 import java.util.function.Predicate;
 import javax.annotation.Nullable;
+
+import org.vivecraft.api.NetworkHelper;
+import org.vivecraft.api.ServerVivePlayer;
+import org.vivecraft.utils.math.Vector3;
+
 import net.minecraft.advancements.CriteriaTriggers;
 import net.minecraft.client.util.ITooltipFlag;
 import net.minecraft.enchantment.EnchantmentHelper;
@@ -259,6 +264,14 @@
                 Vec3d vec3d1 = shooter.getUpVector(1.0F);
                 Quaternion quaternion = new Quaternion(new Vec3f(vec3d1), projectileAngle, true);
                 Vec3d vec3d = shooter.getLook(1.0F);
+            	//VIVE SET AIM AND SPEED
+            	ServerVivePlayer vp = NetworkHelper.vivePlayers.get(shooter.getUniqueID());
+            	if (vp !=null && vp.isVR())
+            	{ 
+            		vec3d = vp.getControllerDir(vp.activeHand);	
+            		vec3d1 = vp.getControllerVectorCustom(vp.activeHand, new Vector3(0, 1, 0));
+            	}
+            	///
                 Vec3f vec3f = new Vec3f(vec3d);
                 vec3f.transform(quaternion);
                 projectileentity.shoot((double)vec3f.getX(), (double)vec3f.getY(), (double)vec3f.getZ(), velocity, inaccuracy);
@@ -417,7 +430,7 @@
         }
     }
 
-    private static float getCharge(int useTime, ItemStack stack)
+    public static float getCharge(int useTime, ItemStack stack)
     {
         float f = (float)useTime / (float)getChargeTime(stack);
 
