--- a/net/minecraft/item/Item.java
+++ b/net/minecraft/item/Item.java
@@ -8,8 +8,14 @@
 import java.util.Random;
 import java.util.UUID;
 import javax.annotation.Nullable;
+
+import org.vivecraft.api.VRData;
+import org.vivecraft.asm.ASMDelegator;
+
 import net.minecraft.block.Block;
 import net.minecraft.block.BlockState;
+import net.minecraft.client.Minecraft;
+import net.minecraft.client.entity.player.ClientPlayerEntity;
 import net.minecraft.client.util.ITooltipFlag;
 import net.minecraft.entity.Entity;
 import net.minecraft.entity.LivingEntity;
@@ -126,7 +132,7 @@
         {
             ItemStack itemstack = playerIn.getHeldItem(handIn);
 
-            if (playerIn.canEat(this.getFood().canEatWhenFull()))
+            if (playerIn.canEat(this.getFood().canEatWhenFull()) || itemstack.getDisplayName().getString().equals("EAT ME"))
             {
                 playerIn.setActiveHand(handIn);
                 return ActionResult.resultConsume(itemstack);
@@ -382,6 +388,13 @@
         float f = player.rotationPitch;
         float f1 = player.rotationYaw;
         Vector3d vector3d = player.getEyePosition(1.0F);
+        
+        //Vivecraft - Identical code to asm for vanilla and so we understand what's going on.
+    	f = ASMDelegator.itemRayTracePitch(player, f);
+    	f1 = ASMDelegator.itemRayTraceYaw(player, f1);
+    	vector3d = ASMDelegator.itemRayTracePos(player, vector3d);
+        //
+    	
         float f2 = MathHelper.cos(-f1 * ((float)Math.PI / 180F) - (float)Math.PI);
         float f3 = MathHelper.sin(-f1 * ((float)Math.PI / 180F) - (float)Math.PI);
         float f4 = -MathHelper.cos(-f * ((float)Math.PI / 180F));
