--- a/net/optifine/ConnectedTextures.java
+++ b/net/optifine/ConnectedTextures.java
@@ -2257,7 +2257,7 @@
             }
             catch (FileNotFoundException filenotfoundexception)
             {
-                Config.warn("ConnectedTextures file not found: " + s);
+                Config.warn("ConnectedTextures file not found: " + s + " " + filenotfoundexception.getMessage());
             }
             catch (Exception exception)
             {
