--- a/net/optifine/ConnectedProperties.java
+++ b/net/optifine/ConnectedProperties.java
@@ -481,7 +481,7 @@
 
     public static IProperty getProperty(String key, Collection properties)
     {
-        for (IProperty iproperty : (Set<IProperty>)(Set<?>)properties)
+        for (IProperty iproperty : (Collection<IProperty>)(Collection<?>)properties)
         {
             if (key.equals(iproperty.getName()))
             {
