--- a/net/minecraft/client/renderer/model/BakedQuad.java
+++ b/net/minecraft/client/renderer/model/BakedQuad.java
@@ -116,11 +116,6 @@
         Reflector.callVoid(Reflector.LightUtil_putBakedQuad, p_pipe_1_, this);
     }
 
-    public TextureAtlasSprite getSprite()
-    {
-        return this.sprite;
-    }
-
     private static TextureAtlasSprite getSpriteByUv(int[] p_getSpriteByUv_0_)
     {
         float f = 1.0F;
