--- a/com/mojang/blaze3d/platform/GlStateManager.java
+++ b/com/mojang/blaze3d/platform/GlStateManager.java
@@ -52,7 +52,7 @@
         return new GlStateManager.BooleanState[p_lambda$static$2_0_];
     });
     private static final GlStateManager.ColorMaterialState COLOR_MATERIAL = new GlStateManager.ColorMaterialState();
-    private static final GlStateManager.BlendState BLEND = new GlStateManager.BlendState();
+    public static final GlStateManager.BlendState BLEND = new GlStateManager.BlendState();
     private static final GlStateManager.DepthState DEPTH = new GlStateManager.DepthState();
     private static final GlStateManager.FogState FOG = new GlStateManager.FogState();
     private static final GlStateManager.CullState CULL = new GlStateManager.CullState();
@@ -332,6 +332,12 @@
     {
         RenderSystem.assertThread(RenderSystem::isOnRenderThread);
 
+        // VIVECRAFT: correct bad blend function that trashes alpha channel
+        if (srcFactor == SourceFactor.SRC_ALPHA.param && dstFactor == DestFactor.ONE_MINUS_SRC_ALPHA.param && srcFactorAlpha == SourceFactor.ONE.param && dstFactorAlpha == DestFactor.ZERO.param) {
+            dstFactorAlpha = DestFactor.ONE_MINUS_SRC_ALPHA.param;
+        }
+        //
+
         if (blendLock.isLocked())
         {
             blendLockState.setFactors(srcFactor, dstFactor, srcFactorAlpha, dstFactorAlpha);
@@ -1011,7 +1017,13 @@
         getMatrix(2982, MATRIX_BUFFER);
         multMatrix(MATRIX_BUFFER);
     }
-
+    
+    public static void mulTextureByModelView()
+    {
+        getMatrix(2982, MATRIX_BUFFER);
+        multMatrix(MATRIX_BUFFER);
+    }
+    
     @Deprecated
     public static void enableFog()
     {
@@ -1537,7 +1549,18 @@
         RenderSystem.assertThread(RenderSystem::isOnRenderThread);
         GL11.glGetFloatv(pname, params);
     }
-
+    
+    //temp
+    public static Matrix4f getMatrix4f(int p_getMatrix4f_0_)
+    {
+        GL11.glGetFloatv(p_getMatrix4f_0_, MATRIX_BUFFER);
+        MATRIX_BUFFER.rewind();
+        Matrix4f matrix4f = new Matrix4f();
+        matrix4f.read(MATRIX_BUFFER);
+        MATRIX_BUFFER.rewind();
+        return matrix4f;
+    }
+    
     @Deprecated
     public static void ortho(double left, double right, double bottom, double top, double zNear, double zFar)
     {
@@ -2051,7 +2074,7 @@
         }
     }
 
-    static class BlendState
+    public static class BlendState
     {
         public final GlStateManager.BooleanState blend = new GlStateManager.BooleanState(3042);
         public int srcFactorRgb = 1;
@@ -2105,6 +2128,7 @@
     }
 
     @Deprecated
+	public
     static class Color
     {
         public float red = 1.0F;
