--- a/net/minecraft/client/gui/AbstractGui.java
+++ b/net/minecraft/client/gui/AbstractGui.java
@@ -181,6 +181,11 @@
 
     private static void innerBlit(Matrix4f p_238461_0_, int p_238461_1_, int p_238461_2_, int p_238461_3_, int p_238461_4_, int p_238461_5_, float p_238461_6_, float p_238461_7_, float p_238461_8_, float p_238461_9_)
     {
+        // VIVECRAFT: DestFactor of ONE is really what we want here, because math
+        // Also this blending wasn't here before but it fixes stuff so yeah
+    	RenderSystem.enableBlend();
+    	RenderSystem.blendFuncSeparate(GlStateManager.SourceFactor.SRC_ALPHA, GlStateManager.DestFactor.ONE_MINUS_SRC_ALPHA, GlStateManager.SourceFactor.ONE, GlStateManager.DestFactor.ONE);
+    	//
         BufferBuilder bufferbuilder = Tessellator.getInstance().getBuffer();
         bufferbuilder.begin(7, DefaultVertexFormats.POSITION_TEX);
         bufferbuilder.pos(p_238461_0_, (float)p_238461_1_, (float)p_238461_4_, (float)p_238461_5_).tex(p_238461_6_, p_238461_9_).endVertex();
@@ -190,6 +195,9 @@
         bufferbuilder.finishDrawing();
         RenderSystem.enableAlphaTest();
         WorldVertexBufferUploader.draw(bufferbuilder);
+        //
+        RenderSystem.disableBlend();
+        //
     }
 
     public int getBlitOffset()
