--- a/net/minecraft/client/gui/INestedGuiEventHandler.java
+++ b/net/minecraft/client/gui/INestedGuiEventHandler.java
@@ -7,6 +7,8 @@
 import java.util.function.Supplier;
 import javax.annotation.Nullable;
 
+import net.minecraft.client.Minecraft;
+
 public interface INestedGuiEventHandler extends IGuiEventListener
 {
     List <? extends IGuiEventListener > getEventListeners();
@@ -91,10 +93,14 @@
     void setListener(@Nullable IGuiEventListener listener);
 
 default void setFocusedDefault(@Nullable IGuiEventListener eventListener)
-    {
-        this.setListener(eventListener);
-        eventListener.changeFocus(true);
-    }
+{
+	//Vivecraft - dont autofocus text fields, it forces open the keyboard
+	if(Minecraft.getInstance().vrSettings.seated) {
+		this.setListener(eventListener);
+		eventListener.changeFocus(true);
+	}
+	//
+}
 
 default void setListenerDefault(@Nullable IGuiEventListener eventListener)
     {
