--- a/net/minecraft/client/gui/IngameGui.java
+++ b/net/minecraft/client/gui/IngameGui.java
@@ -6,8 +6,11 @@
 import com.google.common.collect.Ordering;
 import com.mojang.blaze3d.matrix.MatrixStack;
 import com.mojang.blaze3d.platform.GlStateManager;
+import com.mojang.blaze3d.platform.GlStateManager.DestFactor;
+import com.mojang.blaze3d.platform.GlStateManager.SourceFactor;
 import com.mojang.blaze3d.systems.RenderSystem;
 import com.mojang.datafixers.util.Pair;
+
 import java.util.Collection;
 import java.util.Iterator;
 import java.util.List;
@@ -16,6 +19,9 @@
 import java.util.UUID;
 import java.util.stream.Collectors;
 import javax.annotation.Nullable;
+
+import org.lwjgl.opengl.GL11;
+
 import net.minecraft.block.Blocks;
 import net.minecraft.client.GameSettings;
 import net.minecraft.client.Minecraft;
@@ -27,11 +33,13 @@
 import net.minecraft.client.gui.overlay.DebugOverlayGui;
 import net.minecraft.client.gui.overlay.PlayerTabOverlayGui;
 import net.minecraft.client.gui.overlay.SubtitleOverlayGui;
+import net.minecraft.client.gui.screen.Screen;
 import net.minecraft.client.gui.screen.inventory.ContainerScreen;
 import net.minecraft.client.renderer.ActiveRenderInfo;
 import net.minecraft.client.renderer.BufferBuilder;
 import net.minecraft.client.renderer.ItemRenderer;
 import net.minecraft.client.renderer.Tessellator;
+import net.minecraft.client.renderer.WorldVertexBufferUploader;
 import net.minecraft.client.renderer.texture.AtlasTexture;
 import net.minecraft.client.renderer.texture.PotionSpriteUploader;
 import net.minecraft.client.renderer.texture.TextureAtlasSprite;
@@ -44,6 +52,7 @@
 import net.minecraft.entity.player.PlayerEntity;
 import net.minecraft.inventory.container.INamedContainerProvider;
 import net.minecraft.item.ItemStack;
+import net.minecraft.item.Items;
 import net.minecraft.potion.Effect;
 import net.minecraft.potion.EffectInstance;
 import net.minecraft.potion.Effects;
@@ -67,6 +76,7 @@
 import net.minecraft.util.text.IFormattableTextComponent;
 import net.minecraft.util.text.ITextComponent;
 import net.minecraft.util.text.StringTextComponent;
+import net.minecraft.util.text.TextComponent;
 import net.minecraft.util.text.TextFormatting;
 import net.minecraft.util.text.TranslationTextComponent;
 import net.minecraft.world.GameType;
@@ -115,6 +125,8 @@
     private int scaledWidth;
     private int scaledHeight;
     private final Map<ChatType, List<IChatListener>> chatListeners = Maps.newHashMap();
+    // Vivecraft
+    public boolean showPlayerList;
 
     public IngameGui(Minecraft p_i32_1_)
     {
@@ -155,32 +167,40 @@
         FontRenderer fontrenderer = this.getFontRenderer();
         RenderSystem.enableBlend();
 
-        if (Config.isVignetteEnabled())
-        {
-            this.renderVignette(this.mc.getRenderViewEntity());
-        }
-        else
-        {
+        //Vivecraft no no
+//        if (Config.isVignetteEnabled())
+//        {
+//            this.renderVignette(this.mc.getRenderViewEntity());
+//        }
+//        else
+//        {
+//        	
+//        }
+
             RenderSystem.enableDepthTest();
-            RenderSystem.defaultBlendFunc();
-        }
+      //''  RenderSystem.defaultBlendFunc();
+        RenderSystem.blendFuncSeparate(GlStateManager.SourceFactor.SRC_ALPHA, GlStateManager.DestFactor.ONE_MINUS_SRC_ALPHA, GlStateManager.SourceFactor.ONE, GlStateManager.DestFactor.ONE);
+
 
         ItemStack itemstack = this.mc.player.inventory.armorItemInSlot(3);
 
         if (this.mc.gameSettings.thirdPersonView == 0 && itemstack.getItem() == Blocks.CARVED_PUMPKIN.asItem())
         {
-            this.renderPumpkinOverlay();
+            //this.renderPumpkinOverlay();
         }
+        //
 
-        if (!this.mc.player.isPotionActive(Effects.NAUSEA))
-        {
-            float f = MathHelper.lerp(p_238445_2_, this.mc.player.prevTimeInPortal, this.mc.player.timeInPortal);
 
-            if (f > 0.0F)
-            {
-                this.renderPortal(f);
-            }
-        }
+        //Vivecraft oh no no no.
+//        if (!this.mc.player.isPotionActive(Effects.NAUSEA))
+//        {
+//            float f = MathHelper.lerp(p_238445_2_, this.mc.player.prevTimeInPortal, this.mc.player.timeInPortal);
+
+//            if (f > 0.0F)
+//            {
+//                this.renderPortal(f);
+//            }
+//        }
 
         if (this.mc.playerController.getCurrentGameType() == GameType.SPECTATOR)
         {
@@ -197,9 +217,13 @@
             this.mc.getTextureManager().bindTexture(GUI_ICONS_LOCATION);
             RenderSystem.enableBlend();
             RenderSystem.enableAlphaTest();
-            this.renderAttackIndicator(p_238445_1_);
+            //Vivecraft nope
+            //this.renderAttackIndicator(p_238445_1_);
+            //
             GlStateManager.enableAlphaTest();
-            RenderSystem.defaultBlendFunc();
+            //Vivecraft blend
+            RenderSystem.blendFuncSeparate(GlStateManager.SourceFactor.SRC_ALPHA, GlStateManager.DestFactor.ONE_MINUS_SRC_ALPHA, GlStateManager.SourceFactor.ONE, GlStateManager.DestFactor.ONE);
+            ///
             this.mc.getProfiler().startSection("bossHealth");
             this.overlayBoss.func_238484_a_(p_238445_1_);
             this.mc.getProfiler().endSection();
@@ -285,7 +309,9 @@
                     RenderSystem.pushMatrix();
                     RenderSystem.translatef((float)(this.scaledWidth / 2), (float)(this.scaledHeight - 68), 0.0F);
                     RenderSystem.enableBlend();
-                    RenderSystem.defaultBlendFunc();
+                    //Vivecraft blend
+                    RenderSystem.blendFuncSeparate(GlStateManager.SourceFactor.SRC_ALPHA, GlStateManager.DestFactor.ONE_MINUS_SRC_ALPHA, GlStateManager.SourceFactor.ONE, GlStateManager.DestFactor.ONE);
+                    ///
                     int k1 = 16777215;
 
                     if (this.animateOverlayMessageColor)
@@ -328,7 +354,9 @@
                     RenderSystem.pushMatrix();
                     RenderSystem.translatef((float)(this.scaledWidth / 2), (float)(this.scaledHeight / 2), 0.0F);
                     RenderSystem.enableBlend();
-                    RenderSystem.defaultBlendFunc();
+                    //Vivecraft blend
+                    RenderSystem.blendFuncSeparate(GlStateManager.SourceFactor.SRC_ALPHA, GlStateManager.DestFactor.ONE_MINUS_SRC_ALPHA, GlStateManager.SourceFactor.ONE, GlStateManager.DestFactor.ONE);
+                    ///
                     RenderSystem.pushMatrix();
                     RenderSystem.scalef(4.0F, 4.0F, 4.0F);
                     int l1 = j1 << 24 & -16777216;
@@ -375,9 +403,30 @@
             {
                 this.renderScoreboard(p_238445_1_, scoreobjective1);
             }
+            
+			if (mc.currentScreen != null) {
+				// VIVECRAFT: Dimming effect when in background
+				float brightness = 0.5F;
+				GlStateManager.enableBlend();
+				RenderSystem.blendFuncSeparate(SourceFactor.ZERO, DestFactor.ONE, SourceFactor.ZERO, DestFactor.SRC_ALPHA);
+				GlStateManager.disableTexture();
+				GlStateManager.depthMask(false);
+				Tessellator tessellator = Tessellator.getInstance();
+				BufferBuilder bufferbuilder = tessellator.getBuffer();
+				bufferbuilder.begin(GL11.GL_QUADS, DefaultVertexFormats.POSITION_COLOR);
+				bufferbuilder.pos(0.0D, mc.currentScreen.height, 990.0D).color(0.0F, 0.0F, 0.0F, brightness).endVertex();
+				bufferbuilder.pos(mc.currentScreen.width, mc.currentScreen.height, 990.0D).color(0.0F, 0.0F, 0.0F, brightness).endVertex();
+				bufferbuilder.pos(mc.currentScreen.width, 0.0D, 990.0D).color(0.0F, 0.0F, 0.0F, brightness).endVertex();
+				bufferbuilder.pos(0.0D, 0.0D, 990.0D).color(0.0F, 0.0F, 0.0F, brightness).endVertex();
+				tessellator.draw();
+				GlStateManager.depthMask(true);
+				GlStateManager.enableTexture();
+			}
 
             RenderSystem.enableBlend();
-            RenderSystem.defaultBlendFunc();
+            //Vivecraft blend
+            RenderSystem.blendFuncSeparate(GlStateManager.SourceFactor.SRC_ALPHA, GlStateManager.DestFactor.ONE_MINUS_SRC_ALPHA, GlStateManager.SourceFactor.ONE, GlStateManager.DestFactor.ONE);
+            ///
             RenderSystem.disableAlphaTest();
             RenderSystem.pushMatrix();
             RenderSystem.translatef(0.0F, (float)(this.scaledHeight - 48), 0.0F);
@@ -387,7 +436,7 @@
             RenderSystem.popMatrix();
             scoreobjective1 = scoreboard.getObjectiveInDisplaySlot(0);
 
-            if (this.mc.gameSettings.keyBindPlayerList.isKeyDown() && (!this.mc.isIntegratedServerRunning() || this.mc.player.connection.getPlayerInfoMap().size() > 1 || scoreobjective1 != null))
+            if ((this.mc.gameSettings.keyBindPlayerList.isKeyDown() || showPlayerList) && (!this.mc.isIntegratedServerRunning() || this.mc.player.connection.getPlayerInfoMap().size() > 1 || scoreobjective1 != null))
             {
                 this.overlayPlayerList.setVisible(true);
                 this.overlayPlayerList.func_238523_a_(p_238445_1_, this.scaledWidth, scoreboard, scoreobjective1);
@@ -582,7 +631,7 @@
 
                     if (Reflector.IForgeEffectInstance_renderHUDEffect.exists())
                     {
-                        Reflector.call(effectinstance, Reflector.IForgeEffectInstance_renderHUDEffect, this, k, l, this.getBlitOffset(), f);
+                        Reflector.call(effectinstance, Reflector.IForgeEffectInstance_renderHUDEffect, this, p_238444_1_, k, l, this.getBlitOffset(), f);
                     }
                 }
             }
@@ -622,7 +671,9 @@
             this.setBlitOffset(j);
             RenderSystem.enableRescaleNormal();
             RenderSystem.enableBlend();
-            RenderSystem.defaultBlendFunc();
+            //Vivecraft blend
+            RenderSystem.blendFuncSeparate(GlStateManager.SourceFactor.SRC_ALPHA, GlStateManager.DestFactor.ONE_MINUS_SRC_ALPHA, GlStateManager.SourceFactor.ONE, GlStateManager.DestFactor.ONE);
+            //
             CustomItems.setRenderOffHand(false);
 
             for (int i1 = 0; i1 < 9; ++i1)
@@ -673,6 +724,10 @@
 
             RenderSystem.disableRescaleNormal();
             RenderSystem.disableBlend();
+
+            //Vivecraft
+            renderViveHudIcons(p_238443_2_);
+            //
         }
     }
 
@@ -1212,6 +1267,7 @@
         }
     }
 
+    //unused
     private void renderPumpkinOverlay()
     {
         RenderSystem.disableDepthTest();
@@ -1243,6 +1299,8 @@
         }
     }
 
+    
+    //unused
     private void renderVignette(Entity entityIn)
     {
         if (!Config.isVignetteEnabled())
@@ -1295,6 +1353,7 @@
         }
     }
 
+    //unused
     private void renderPortal(float timeInPortal)
     {
         if (timeInPortal < 1.0F)
@@ -1393,8 +1452,8 @@
 
             if (Reflector.IForgeItemStack_getHighlightTip.exists())
             {
-                String s = Reflector.callString(itemstack, Reflector.IForgeItemStack_getHighlightTip, itemstack.getDisplayName());
-                String s1 = Reflector.callString(this.highlightingItemStack, Reflector.IForgeItemStack_getHighlightTip, this.highlightingItemStack.getDisplayName());
+                ITextComponent s = (ITextComponent) Reflector.call(itemstack, Reflector.IForgeItemStack_getHighlightTip, itemstack.getDisplayName());
+                ITextComponent s1 = (ITextComponent) Reflector.call(this.highlightingItemStack, Reflector.IForgeItemStack_getHighlightTip, this.highlightingItemStack.getDisplayName());
                 flag = Config.equals(s, s1);
             }
 
@@ -1520,4 +1579,80 @@
     {
         this.overlayDebug.resetChunk();
     }
+    
+	//VIVECRAFT ADDITIONS ***********************************
+  	
+  	private void renderViveHudIcons(MatrixStack matrixstack){
+          //VIVE SPRINTDICATOR
+          if (this.mc.getRenderViewEntity() instanceof PlayerEntity)
+          {
+        	  int i = mc.getMainWindow().getScaledWidth();
+        	  int j = mc.getMainWindow().getScaledHeight();
+        	  FontRenderer fontrenderer = mc.ingameGUI.getFontRenderer();
+        	  PlayerEntity entityplayer = (PlayerEntity)this.mc.getRenderViewEntity();
+        	  int iconp = 0;    	  
+        	
+        	  Effect e = null;
+        	  if(entityplayer.isSprinting()) e = Effects.SPEED;      	  
+        	  if(entityplayer.isSneaking()) e = Effects.BLINDNESS;
+        	  if(entityplayer.isActualySwimming()) e = Effects.DOLPHINS_GRACE;
+        	  if(entityplayer.isElytraFlying()) iconp = -1;
+        	  
+        	  if(iconp == -1){
+//            	  GlStateManager.color4f(1.0F, 1.0F, 1.0F, 1.0F);
+        		  int w = this.mc.getMainWindow().getScaledWidth() / 2 - 109;
+        		  int h = this.mc.getMainWindow().getScaledHeight() -39;	 
+//        		  //TODO: this aint right.
+//        		  TextureAtlasSprite textureatlassprite =  this.mc.getAtlasSpriteGetter(AtlasTexture.LOCATION_BLOCKS_TEXTURE).apply(new ResourceLocation("textures/items/elytra"));
+//        		  this.mc.getTextureManager().bindTexture(AtlasTexture.LOCATION_BLOCKS_TEXTURE);
+//        		  AbstractGui.blit(w,h,0,16,16,textureatlassprite);
+//        		  
+        		  mc.getItemRenderer().renderItemIntoGUI(new ItemStack(Items.ELYTRA), w, h);
+        		  
+        	//	    public void renderItemAndEffectIntoGUI(@Nullable LivingEntity entityIn, ItemStack itemIn, int x, int y)
+
+        		  
+        	  } 
+        	  else if (e!=null) {  		   
+        	      TextureAtlasSprite sprite = mc.getPotionSpriteUploader().getSprite(e);
+            	  GlStateManager.color4f(1.0F, 1.0F, 1.0F, 1.0F);
+        		  this.mc.getTextureManager().bindTexture(sprite.getAtlasTexture().getTextureLocation());
+        		  int w = this.mc.getMainWindow().getScaledWidth() / 2 - 109;
+        		  int h = this.mc.getMainWindow().getScaledHeight() -39;      			  
+        		  AbstractGui.blit(matrixstack, w, h, 0, 18, 18, sprite);        			
+        	  }       	  
+          }
+          //
+  	}
+
+	public void drawMouseMenuQuad(int mouseX, int mouseY)
+	{	
+		RenderSystem.enableBlend();
+		RenderSystem.enableAlphaTest();
+		RenderSystem.disableDepthTest();
+		RenderSystem.disableLighting();
+		RenderSystem.color4f(1, 1, 1, 1);
+
+		this.mc.getTextureManager().bindTexture(GUI_ICONS_LOCATION);
+		float menuMousePointerSize = 16f * this.mc.vrSettings.menuCrosshairScale;
+		
+		// Turns out all we needed was some blendFuncSeparate magic :)
+		// Also color DestFactor of ZERO produces better results with non-white crosshairs
+		RenderSystem.blendFuncSeparate(SourceFactor.ONE_MINUS_DST_COLOR, DestFactor.ZERO, GlStateManager.SourceFactor.ONE, GlStateManager.DestFactor.ONE);
+		drawCentredTexturedModalRect(mouseX, mouseY, menuMousePointerSize, menuMousePointerSize, 0, 0, 15, 15);
+	}
+	
+	public void drawCentredTexturedModalRect(int centreX, int centreY, float width, float height, int u, int v, int texWidth, int texHeight)
+	{   	
+		float f = 1f/256;
+		float f1 = 1f/256;
+		BufferBuilder bufferbuilder = Tessellator.getInstance().getBuffer();
+		bufferbuilder.begin(GL11.GL_QUADS, DefaultVertexFormats.POSITION_TEX);
+		bufferbuilder.pos((double)(centreX - (width / 2f)), (double)(centreY + (height / 2f)), (double)this.getBlitOffset()).tex(((float)(u + 0) * f), ((float)(v + texHeight) * f1)).endVertex();
+		bufferbuilder.pos((double)(centreX + (width / 2f)), (double)(centreY + (height / 2f)), (double)this.getBlitOffset()).tex( ((float)(u + texWidth) * f), ((float)(v + texHeight) * f1)).endVertex();
+		bufferbuilder.pos((double)(centreX + (width / 2f)), (double)(centreY - (height / 2f)), (double)this.getBlitOffset()).tex( ((float)(u + texWidth) * f), ((float)(v + 0) * f1)).endVertex();
+		bufferbuilder.pos((double)(centreX - (width / 2f)), (double)(centreY - (height / 2f)), (double)this.getBlitOffset()).tex( ((float)(u + 0) * f), ((float)(v + 0) * f1)).endVertex();
+	    bufferbuilder.finishDrawing();
+	    WorldVertexBufferUploader.draw(bufferbuilder);
+	}
 }
