--- a/net/minecraft/client/gui/NewChatGui.java
+++ b/net/minecraft/client/gui/NewChatGui.java
@@ -166,7 +166,7 @@
 
     public void printChatMessage(ITextComponent chatComponent)
     {
-        this.printChatMessageWithOptionalDeletion(chatComponent, 0);
+    	this.printChatMessageWithOptionalDeletion(chatComponent, 0);
     }
 
     public void printChatMessageWithOptionalDeletion(ITextComponent chatComponent, int chatLineId)
